{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#%%appyter init\n",
    "from appyter import magic\n",
    "magic.init(lambda _=globals: _())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter hide_code\n",
    "\n",
    "{% do SectionField(\n",
    "    name='primary',\n",
    "    title='Upload Tumor Expression',\n",
    "    img='upload.png'\n",
    ") %}\n",
    "\n",
    "\n",
    "{% set rna_file = CustomFileField(\n",
    "    name='rna_expr',\n",
    "    label='RNA-seq expression',\n",
    "    description='''\n",
    "    File should be a tsv/csv of the form:\n",
    "\n",
    "        Patient 1 Tumor    Patient 2 Tumor  ...\n",
    "    ---------------------------------------\n",
    "    Gene 1    0                      200                ...\n",
    "    ---------------------------------------\n",
    "    Gene 2    5                      180                ...\n",
    "    ---------------------------------------\n",
    "      ...           ...                      ...                  ...\n",
    "    ''',\n",
    "    default='LUAD_tumor_counts.tsv',\n",
    "    required=False,\n",
    "    examples={\n",
    "        'BR_tumor_counts.tsv': 'https://minio.dev.maayanlab.cloud/x2k-tr/BR_tumor_counts.tsv',\n",
    "        'CCRCC_tumor_counts.tsv': 'https://minio.dev.maayanlab.cloud/x2k-tr/CCRCC_tumor_counts.tsv',\n",
    "        'CO_tumor_counts.tsv': 'https://minio.dev.maayanlab.cloud/x2k-tr/CO_tumor_counts.tsv',\n",
    "        'GBM_tumor_counts.tsv': 'https://minio.dev.maayanlab.cloud/x2k-tr/GBM_tumor_counts.tsv',\n",
    "        'HNSCC_tumor_counts.tsv': 'https://minio.dev.maayanlab.cloud/x2k-tr/HNSCC_tumor_counts.tsv',\n",
    "        'LSCC_tumor_counts.tsv': 'https://minio.dev.maayanlab.cloud/x2k-tr/LSCC_tumor_counts.tsv',\n",
    "        'LUAD_tumor_counts.tsv': 'https://minio.dev.maayanlab.cloud/x2k-tr/LUAD_tumor_counts.tsv',\n",
    "        'OV_tumor_counts.tsv': 'https://minio.dev.maayanlab.cloud/x2k-tr/OV_tumor_counts.tsv',\n",
    "        'PDAC_tumor_counts.tsv': 'https://minio.dev.maayanlab.cloud/x2k-tr/PDAC_tumor_counts.tsv',\n",
    "        'UCEC_tumor_counts.tsv': 'https://minio.dev.maayanlab.cloud/x2k-tr/UCEC_tumor_counts.tsv'\n",
    "    },\n",
    "    section='primary',\n",
    ") %}\n",
    "\n",
    "\n",
    "{% set prot_file = CustomFileField(\n",
    "    name='prot_expr',\n",
    "    label='protein expression',\n",
    "    description='''\n",
    "    File should be a tsv/csv of the form:\n",
    "\n",
    "        Patient 1 Tumor    Patient 2 Tumor  ...\n",
    "    ---------------------------------------\n",
    "    Gene/Protein 1    0                   200       ...\n",
    "    ---------------------------------------\n",
    "    Gene/Protein 2    5                   180       ...\n",
    "    ---------------------------------------\n",
    "      ...                       ...                    ...        ...\n",
    "    ''',\n",
    "    default='LUAD_proteomics_tumor.tsv',\n",
    "    required=False,\n",
    "    examples={\n",
    "        'BR_proteomics_tumor.tsv': 'https://minio.dev.maayanlab.cloud/x2k-tr/BR_proteomics_tumor.tsv',\n",
    "        'CCRCC_proteomics_tumor.tsv': 'https://minio.dev.maayanlab.cloud/x2k-tr/CCRCC_proteomics_tumor.tsv',\n",
    "        'CO_proteomics_tumor.tsv': 'https://minio.dev.maayanlab.cloud/x2k-tr/CO_proteomics_tumor.tsv',\n",
    "        'GBM_proteomics_tumor.tsv': 'https://minio.dev.maayanlab.cloud/x2k-tr/GBM_proteomics_tumor.tsv',\n",
    "        'HNSCC_proteomics_tumor.tsv': 'https://minio.dev.maayanlab.cloud/x2k-tr/HNSCC_proteomics_tumor.tsv',\n",
    "        'LSCC_proteomics_tumor.tsv': 'https://minio.dev.maayanlab.cloud/x2k-tr/LSCC_proteomics_tumor.tsv',\n",
    "        'LUAD_proteomics_tumor.tsv': 'https://minio.dev.maayanlab.cloud/x2k-tr/LUAD_proteomics_tumor.tsv',\n",
    "        'OV_proteomics_tumor.tsv': 'https://minio.dev.maayanlab.cloud/x2k-tr/OV_proteomics_tumor.tsv',\n",
    "        'PDAC_proteomics_tumor.tsv': 'https://minio.dev.maayanlab.cloud/x2k-tr/PDAC_proteomics_tumor.tsv',\n",
    "        'UCEC_proteomics_tumor.tsv': 'https://minio.dev.maayanlab.cloud/x2k-tr/UCEC_proteomics_tumor.tsv'\n",
    "    },\n",
    "    section='primary',\n",
    ") %}\n",
    "\n",
    "{% set phospho_file = CustomFileField(\n",
    "    name='phospho_expr',\n",
    "    label='protein phosphorylation',\n",
    "    description='''\n",
    "    File should be a tsv/csv of the form:\n",
    "\n",
    "        Patient 1 Tumor    Patient 2 Tumor  ...\n",
    "    ---------------------------------------\n",
    "    Phosphoprotein 1     0        200          ...\n",
    "    ---------------------------------------\n",
    "    Phosphoprotein 2    5         180           ...\n",
    "    ---------------------------------------\n",
    "      ...                          ...              ...          ...\n",
    "    ''',\n",
    "    default='LUAD_phospho_observed.tsv',\n",
    "    required=False,\n",
    "    examples={\n",
    "        'BR_phospho_observed.tsv': 'https://minio.dev.maayanlab.cloud/x2k-tr/BR_phospho_observed.tsv',\n",
    "        'CCRCC_phospho_observed.tsv': 'https://minio.dev.maayanlab.cloud/x2k-tr/CCRCC_phospho_observed.tsv',\n",
    "        'CO_phospho_observed.tsv': 'https://minio.dev.maayanlab.cloud/x2k-tr/CO_phospho_observed.tsv',\n",
    "        'GBM_phospho_observed.tsv': 'https://minio.dev.maayanlab.cloud/x2k-tr/GBM_phospho_observed.tsv',\n",
    "        'HNSCC_phospho_observed.tsv': 'https://minio.dev.maayanlab.cloud/x2k-tr/HNSCC_phospho_observed.tsv',\n",
    "        'LSCC_phospho_observed.tsv': 'https://minio.dev.maayanlab.cloud/x2k-tr/LSCC_phospho_observed.tsv',\n",
    "        'LUAD_phospho_observed.tsv': 'https://minio.dev.maayanlab.cloud/x2k-tr/LUAD_phospho_observed.tsv',\n",
    "        'OV_phospho_observed.tsv': 'https://minio.dev.maayanlab.cloud/x2k-tr/OV_phospho_observed.tsv',\n",
    "        'PDAC_phospho_observed.tsv': 'https://minio.dev.maayanlab.cloud/x2k-tr/PDAC_phospho_observed.tsv',\n",
    "        'UCEC_phospho_observed.tsv': 'https://minio.dev.maayanlab.cloud/x2k-tr/UCEC_phospho_observed.tsv',\n",
    "        \n",
    "    },\n",
    "    section='primary',\n",
    ") %}\n",
    "\n",
    "{% set meta_file = CustomFileField(\n",
    "    name='meta',\n",
    "    label='sample metadata',\n",
    "    description='Samples on the rows, metadata on the columns',\n",
    "    default='LUAD_meta.txt',\n",
    "    required=True,\n",
    "    examples={\n",
    "        'BR_meta.txt': 'https://minio.dev.maayanlab.cloud/x2k-tr/BR_meta.txt',\n",
    "        'CCRCC_meta.txt': 'https://minio.dev.maayanlab.cloud/x2k-tr/CCRCC_meta.txt',\n",
    "        'CO_meta.txt': 'https://minio.dev.maayanlab.cloud/x2k-tr/CO_meta.txt',\n",
    "        'GBM_meta.txt': 'https://minio.dev.maayanlab.cloud/x2k-tr/GBM_meta.txt',\n",
    "        'HNSCC_meta.txt': 'https://minio.dev.maayanlab.cloud/x2k-tr/HNSCC_meta.txt',\n",
    "        'LSCC_meta.txt': 'https://minio.dev.maayanlab.cloud/x2k-tr/LSCC_meta.txt',\n",
    "        'LUAD_meta.txt': 'https://minio.dev.maayanlab.cloud/x2k-tr/LUAD_meta.txt',\n",
    "        'OV_meta.txt': 'https://minio.dev.maayanlab.cloud/x2k-tr/OV_meta.txt',\n",
    "        'PDAC_meta.txt': 'https://minio.dev.maayanlab.cloud/x2k-tr/PDAC_meta.txt',\n",
    "        'UCEC_meta.txt': 'https://minio.dev.maayanlab.cloud/x2k-tr/UCEC_meta.txt'\n",
    "    },\n",
    "    section='primary',\n",
    ") %}\n",
    "\n",
    "{% set membrane_screener_ = BoolField(\n",
    "    name='membrane_screener',\n",
    "    label='Prioritize membrane genes',\n",
    "    description='Use a list of genes that encode cell surface proteins to identify targets that might be suitable for immunotherapy.',\n",
    "    default=True,\n",
    "    section='primary',\n",
    ") %}\n",
    "\n",
    "{% set normalize_to_bg = BoolField(\n",
    "    name='normalize_to_bg',\n",
    "    label='Normalize to background',\n",
    "    description='Normalize the uploaded RNA-seq expression values to the background expression values for target identification through differential gene expression analysis',\n",
    "    default=True,\n",
    "    section='primary',\n",
    ") %}\n",
    "\n",
    "{% set impute_protein_expr = BoolField(\n",
    "    name='impute_protein_expr',\n",
    "    label='Impute protein expression',\n",
    "    description='Impute missing protein expression values with average of non-missing values in rows with at least 80% non-missing values',\n",
    "    default=True,\n",
    "    section='primary',\n",
    ") %}\n",
    "\n",
    "{% set impute_phospho_expr = BoolField(\n",
    "    name='impute_phospho_expr',\n",
    "    label='Impute protein phosphorylation',\n",
    "    description='Impute missing protein phosphorylation values with average of non-missing values in rows with at least 80% non-missing values',\n",
    "    default=True,\n",
    "    section='primary',\n",
    ") %}\n",
    "\n",
    "{% set percent_rows_protein = FloatField(\n",
    "    name='percent_rows_protein',\n",
    "    label='Protein imputation threshold', \n",
    "    default=0.8, \n",
    "    min=0.0,\n",
    "    max=1,\n",
    "    step=0.01,\n",
    "    description='Rows without at least this ratio of non-null values will be dropped for imputation', \n",
    "    section='primary'\n",
    ")\n",
    "%}\n",
    "\n",
    "{% set percent_rows_phospho = FloatField(\n",
    "    name='percent_rows_phospho',\n",
    "    label='Phosphoprotein imputation threshold', \n",
    "    default=0.5, \n",
    "    min=0.0,\n",
    "    max=1,\n",
    "    step=0.01,\n",
    "    description='Rows without at least this ratio of non-null values will be dropped for imputation', \n",
    "    section='primary'\n",
    ")\n",
    "%}\n",
    "\n",
    "\n",
    "{% set chea3_n = IntField(\n",
    "    name='chea3_n',\n",
    "    label='DEGs for ChEA3 Enrichment', \n",
    "    default=500, \n",
    "    min=10,\n",
    "    max=2000,\n",
    "    step=1,\n",
    "    description='Number of genes per patient to be submitted to ChEA3 for enrichment analysis as part of the X2K pipeline.', \n",
    "    section='primary'\n",
    ")\n",
    "%}\n",
    "\n",
    "{% set kea3_n = IntField(\n",
    "    name='kea3_n',\n",
    "    label='DEPPs for KEA3 Enrichment', \n",
    "    default=500, \n",
    "    min=10,\n",
    "    max=2000,\n",
    "    step=1,\n",
    "    description='Number of differentially phosphorylated proteins per patient to be submitted to KEA3 for enrichment analysis.', \n",
    "    section='primary'\n",
    ")\n",
    "%}\n",
    "\n",
    "{% set n_neighbors = IntField(\n",
    "    name='n_neighbors',\n",
    "    label='N neighbors', \n",
    "    default=15, \n",
    "    min=2,\n",
    "    max=100,\n",
    "    step=1,\n",
    "    description='The size of local neighborhood (in terms of number of neighboring data points) used for manifold approximation.', \n",
    "    section='primary'\n",
    ")\n",
    "%}\n",
    "\n",
    "{% set min_dist = FloatField(\n",
    "    name='min_dist',\n",
    "    label='Min Distance', \n",
    "    default=0.01, \n",
    "    min=0.00001,\n",
    "    max=0.9,\n",
    "    step=0.00001,\n",
    "    description='The effective minimum distance between embedded points.', \n",
    "    section='primary'\n",
    ")\n",
    "%}\n",
    "\n",
    "{% set resolution = FloatField(\n",
    "    name='resolution',\n",
    "    label='Resolution', \n",
    "    default=1, \n",
    "    min=0.1,\n",
    "    max=2,\n",
    "    step=0.1,\n",
    "    description='A parameter value controlling the coarseness of the clustering. Higher values lead to more clusters.', \n",
    "    section='primary'\n",
    ")\n",
    "%}\n",
    "\n",
    "{% set tumor_validate = MultiChoiceField(\n",
    "    name='tumor_validate',\n",
    "    label='DepMap cell lines for target validation', \n",
    "    default=[], \n",
    "    description='If using a cancer dataset, choose the tumor type(s) in which you would like to validate the predicted kinases and gene targets. The appropriate DepMap cell lines will be used to quantify the effect of gene knock-down on cancer cell viability.', \n",
    "    choices=['Acute Myeloid Leukemia', 'Adenosquamous Carcinoma of the Pancreas', 'Ampullary Carcinoma', 'Anaplastic Thyroid Cancer', 'B-Lymphoblastic Leukemia/Lymphoma', 'Bladder Squamous Cell Carcinoma', 'Bladder Urothelial Carcinoma', 'Breast Ductal Carcinoma In Situ', 'Breast Neoplasm, NOS', 'Cervical Adenocarcinoma', 'Cervical Squamous Cell Carcinoma', 'Chondrosarcoma', 'Colorectal Adenocarcinoma', 'Cutaneous Squamous Cell Carcinoma', 'Diffuse Glioma', 'Embryonal Tumor', 'Endometrial Carcinoma', 'Epithelioid Sarcoma', 'Esophageal Squamous Cell Carcinoma', 'Esophagogastric Adenocarcinoma', 'Ewing Sarcoma', 'Extra Gonadal Germ Cell Tumor', 'Fibrosarcoma', 'Gestational Trophoblastic Disease', 'Glassy Cell Carcinoma of the Cervix', 'Head and Neck Carcinoma, Other', 'Head and Neck Squamous Cell Carcinoma', 'Hepatoblastoma', 'Hepatocellular Carcinoma', 'Hepatocellular Carcinoma plus Intrahepatic Cholangiocarcinoma', 'Hodgkin Lymphoma', 'Intracholecystic Papillary Neoplasm', 'Intraductal Papillary Neoplasm of the Bile Duct', 'Invasive Breast Carcinoma', 'Leiomyosarcoma', 'Liposarcoma', 'Lung Neuroendocrine Tumor', 'Mature B-Cell Neoplasms', 'Mature T and NK Neoplasms', 'Melanoma', 'Meningothelial Tumor', 'Merkel Cell Carcinoma', 'Mixed Cervical Carcinoma', 'Mucosal Melanoma of the Vulva/Vagina', 'Myelodysplastic Syndromes', 'Myeloproliferative Neoplasms', 'Nerve Sheath Tumor', 'Neuroblastoma', 'Non-Cancerous', 'Non-Hodgkin Lymphoma', 'Non-Seminomatous Germ Cell Tumor', 'Non-Small Cell Lung Cancer', 'Ocular Melanoma', 'Osteosarcoma', 'Ovarian Epithelial Tumor', 'Ovarian Germ Cell Tumor', 'Pancreatic Adenocarcinoma', 'Pleural Mesothelioma', 'Poorly Differentiated Thyroid Cancer', 'Prostate Adenocarcinoma', 'Renal Cell Carcinoma', 'Retinoblastoma', 'Rhabdoid Cancer', 'Rhabdomyosarcoma', 'Salivary Carcinoma', 'Sarcoma, NOS', 'Squamous Cell Carcinoma of the Vulva/Vagina', 'Synovial Sarcoma', 'T-Lymphoblastic Leukemia/Lymphoma', 'Undifferentiated Pleomorphic Sarcoma/Malignant Fibrous Histiocytoma/High-Grade Spindle Cell Sarcoma', 'Urethral Cancer', 'Uterine Sarcoma/Mesenchymal', 'Well-Differentiated Thyroid Cancer'],\n",
    "    section='primary'\n",
    ")\n",
    "%}\n",
    "\n",
    "{% set membrane_screener_list = 'https://appyters.maayanlab.cloud/storage/Tumor_Gene_Target_Screener/surfaceome.csv' %}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from helpers import *\n",
    "import os\n",
    "import re\n",
    "import random\n",
    "import qnorm\n",
    "from string import ascii_uppercase\n",
    "from tqdm import tqdm\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "\n",
    "from matplotlib import cm\n",
    "from matplotlib.patches import Rectangle\n",
    "from matplotlib.patches import Patch\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from matplotlib.colors import LinearSegmentedColormap\n",
    "%matplotlib inline\n",
    "\n",
    "import networkx as nx\n",
    "\n",
    "from IPython.display import HTML, display, Markdown, FileLink\n",
    "from matplotlib.gridspec import GridSpec\n",
    "from matplotlib_venn import venn2\n",
    "from maayanlab_bioinformatics.normalization import zscore_normalize, log2_normalize\n",
    "from scipy.stats import zscore\n",
    "from scipy.stats import ttest_ind, ttest_1samp, ttest_ind_from_stats\n",
    "from maayanlab_bioinformatics.dge import limma_voom_differential_expression\n",
    "from maayanlab_bioinformatics.harmonization.ncbi_genes import ncbi_genes_lookup\n",
    "from maayanlab_bioinformatics.api import enrichr_link_from_genes\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "\n",
    "import warnings\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "os.makedirs('results', exist_ok=True)\n",
    "os.makedirs('figures', exist_ok=True)\n",
    "try:\n",
    "    lookup = ncbi_genes_lookup(organism='Mammalia/Homo_sapiens')\n",
    "except:\n",
    "    import urllib.request, json \n",
    "    with urllib.request.urlopen(\"https://minio.dev.maayanlab.cloud/x2k-tr/ncbi_genes_disambiguated.json\") as url:\n",
    "        ncbi_genes_disambiguated = json.load(url)\n",
    "        lookup = ncbi_genes_disambiguated.get\n",
    "    \n",
    "fig_counter = 2\n",
    "table_counter = 1\n",
    "discussion_results = {}\n",
    "\n",
    "import sys\n",
    "import contextlib\n",
    "@contextlib.contextmanager\n",
    "def suppress_output(stdout=True, stderr=True, dest=os.devnull):\n",
    "    ''' Usage:\n",
    "    with suppress_output():\n",
    "        print('hi')\n",
    "    '''\n",
    "    dev_null = open(dest, 'a')\n",
    "    if stdout:\n",
    "        _stdout = sys.stdout\n",
    "        sys.stdout = dev_null\n",
    "    if stderr:\n",
    "        _stderr = sys.stderr\n",
    "        sys.stderr = dev_null\n",
    "    try:\n",
    "        yield\n",
    "    finally:\n",
    "        if stdout:\n",
    "            sys.stdout = _stdout\n",
    "        if stderr:\n",
    "            sys.stderr = _stderr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "has_rna_file = {{ rna_file }} != ''\n",
    "has_prot_file = {{ prot_file }} != ''\n",
    "has_phospho_file = {{ phospho_file }} != ''\n",
    "has_meta_file = {{ meta_file }} != ''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "if has_rna_file:\n",
    "    rna_df = read_table({{ rna_file }})\n",
    "    rna_filename = {{ rna_file }}.split('.')[0]\n",
    "if has_prot_file:\n",
    "    prot_df = read_table({{ prot_file }})\n",
    "    prot_filename = {{ prot_file }}.split('.')[0]\n",
    "if has_phospho_file:\n",
    "    phospho_df = read_table({{ phospho_file }})\n",
    "    phospho_filename = {{ phospho_file }}.split('.')[0]\n",
    "if has_meta_file:\n",
    "    meta_df = read_table({{ meta_file }})\n",
    "else:\n",
    "    meta_df = pd.DataFrame()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "example_mapper = {\n",
    "    'BR': 'breast invasive carcinoma (BR)', \n",
    "    'CCRCC': 'clear cell renal cell carcinoma (ccRCC)',\n",
    "    'CO': '\tcolon adenocarcinoma (COAD)',\n",
    "    'GBM': 'glioblastoma multiforme (GBM)',\n",
    "    'HNSCC': 'head and neck squamous cell carcinoma (HNSCC)',\n",
    "    'LSCC': 'laryngeal squamous cell carcinoma (LSCC)',\n",
    "    'LUAD': 'lung adenocarcinoma (LUAD)',\n",
    "    'OV': 'ovarian serous cystadenocarcinoma (OV)',\n",
    "    'PDAC': 'pancreatic adenocarcinoma (PDAC)',\n",
    "    'UCEC': 'uterine corpus endometrial carcinoma (UCEC)'\n",
    " }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file and has_phospho_file and has_phospho_file:\n",
    "    example_name = rna_filename.split('_')[0] + ' Samples' if rna_filename.split('_')[0] in example_mapper else ' Cancer'\n",
    "    display(Markdown(f'## Multiomics2Targets: Computational Workflow to Identify Pathways and Targets for {example_name} Profiled with Transcriptomics, Proteomics, and Phosphoproteomics'))\n",
    "elif has_rna_file and has_phospho_file:\n",
    "    example_name = rna_filename.split('_')[0] + ' Samples' if rna_filename.split('_')[0] in example_mapper else ' Cancer'\n",
    "    display(Markdown(f'## Multiomics2Targets: Computational Workflow to Identify Pathways and Targets for {example_name} Profiled with Transcriptomics and Phosphoproteomics'))\n",
    "elif has_rna_file and has_prot_file:\n",
    "    example_name = rna_filename.split('_')[0] + ' Samples' if rna_filename.split('_')[0] in example_mapper else ' Cancer'\n",
    "    display(Markdown(f'## Multiomics2Targets: Computational Workflow to Identify Pathways and Targets for {example_name} Profiled with Transcriptomics and Proteomics'))\n",
    "elif has_phospho_file and has_prot_file:\n",
    "    example_name = phospho_filename.split('_')[0] + ' Samples' if phospho_filename.split('_')[0] in example_mapper else ' Cancer'\n",
    "    display(Markdown(f'## Multiomics2Targets: Computational Workflow to Identify Pathways and Targets for {example_name} Profiled with Phosphoproteomics and Proteomics'))\n",
    "elif has_rna_file:\n",
    "    example_name = rna_filename.split('_')[0] + ' Samples' if rna_filename.split('_')[0] in example_mapper else ' Cancer'\n",
    "    display(Markdown(f'## Multiomics2Targets: Computational Workflow to Identify Pathways and Targets for {example_name} Profiled with Transcriptomics'))\n",
    "elif has_prot_file:\n",
    "    example_name = prot_filename.split('_')[0] + ' Samples' if prot_filename.split('_')[0] in example_mapper else ' Cancer'\n",
    "    display(Markdown(f'## Multiomics2Targets: Computational Workflow to Identify Pathways and Targets for {example_name} Profiled with Proteomics'))\n",
    "elif has_phospho_file:\n",
    "    example_name = phospho_filename.split('_')[0] + ' Samples' if phospho_filename.split('_')[0] in example_mapper else ' Cancer'\n",
    "    display(Markdown(f'## Multiomics2Targets: Computational Workflow to Identify Pathways and Targets for {example_name} Profiled with Phosphoproteomics'))\n",
    "else:\n",
    "    display(Markdown(f'## Multiomics2Targets: Insert Subtitle Here'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "display(Markdown('## Abstract'))\n",
    "abstract_text = \"\"\n",
    "abstract_text += '''A major goal in cancer research is to identify cell signaling pathways and tumor- specific proteins that are targetable with immunotherapies such as antibody drug conjugates (ADCs) and chimeric antigen receptor (CAR) T-cells, as well as kinase inhibitors. Multiomics2Targets is a computational pipeline that identifies proteins, genes, and transcripts that are highly expressed in tumors compared to most healthy human tissues and cell types while prioritizing cell-surface proteins and protein kinases. \n",
    "To identify driver cell signaling pathways and potential immunotherapeutic candidates the pipeline applies the eXpression to Kinases (X2K) [1, 2] and TargetRanger [3] algorithms.'''\n",
    "if has_rna_file:\n",
    "    abstract_text += f\" The {rna_df.shape[1]} {example_mapper.get(rna_filename.split('_')[0], 'tumor')} samples are first clustered into subtypes and then cell signaling pathways are inferred from the RNA-seq data with X2K for each cluster. Then cell surface targets are prioritized for each cluster compared to healthy tissue and cell type expression.\"\n",
    "if has_rna_file and has_prot_file:\n",
    "    abstract_text += ' These targets are subsequently confirmed for high protein expression in the tumor subtypes based on the proteomics profiling of the same tumors.'\n",
    "if has_phospho_file:\n",
    "    abstract_text += ' Differentially phosphorylated proteins are identified in the tumor samples and used to infer kinase activity with KEA3.'\n",
    "if has_phospho_file and has_rna_file:\n",
    "    abstract_text += ' The inferred kinase activity is then used to identify potential kinase targets, comparing them to the driver kinases identified from the transcriptomics through the X2K pipeline.'\n",
    "if has_phospho_file and has_prot_file and has_rna_file:\n",
    "    abstract_text += ' Samples with significant recovery of kinases inferred from the X2K pipeline are compared to the kinases inferred from the phosphoproteomics data are then visualized with the matched proteomics profiling.'\n",
    "abstract_text += ''' With Multiomics2Targets we can identify a multitude of potential driver pathways and novel targets for each cancer subtype. Some of these targets are supported by prior publications while many others are new and warrant further investigation. \n",
    "Overall, the Multiomics2Targets pipeline is a robust computational approach to harness knowledge from multi-omics datasets to identify promising immunotherapeutic and other targets for preclinical cancer research.'''\n",
    "display(Markdown(abstract_text))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "buttons": true
   },
   "outputs": [],
   "source": [
    "# Load the buttons"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Introduction\n",
    "\n",
    "A prominent challenge in cancer immunotherapy and targeted therapy is identifying viable targets and drugs that can selectively kill, or stop the proliferation, of tumor cells without negatively impacting normal cells and tissues. Furthermore, personalized cancer therapy requires that target and drug candidates are tailored to individual or subgroups of tumors. These pathways and targets can be characterized as the molecular alterations of regulatory mechanisms, which are known to vary across different tumor types and individuals [4]. Many computational methods have been developed to elucidate key molecular mechanisms, identify targets and pathways in individual tumors. For example, we developed the eXpression2Kinases (X2K) pipeline [1, 2]. The X2K pipeline takes a set of differentially expressed genes and infers the upstream regulatory network likely responsible for the observed changes in gene expression upstream by chaining the tools ChEA3 [5], Genes2Networks [6], and KEA3 [7]. Other and algorithms for pathway analysis from RNA-seq often use gene expression as a proxy for protein activity. Often such methods are limited to identifying specific curated cell signaling pathways [8, 9]. By contrast, X2K chains several enrichment and network analysis steps to reconstruct an original pathway from information encoded in transcription-factor/target, protein-protein interactions, and kinase-substrate networks. Once such upstream cell signaling pathways have been identified, these can be targeted, for example, by knocking out the top ranked kinases from last step of the X2K workflow [10].\n",
    "\n",
    "\n",
    "Computational approaches have been developed to identify cell surface protein targets for a specific group of cancer patients profiled with RNA-seq with experimental demonstrations that these approaches can work. For example, to identify immunotherapuetic targets for a neuroblastoma patient [11], gene expression data from the patient’s tumor were compared to RNA-seq data from GTEx [12]. Targets were prioritized if they were highly expressed in the tumor compared with their expression across 54 normal GTEx tissues. Genes that met this criterion were further ﬁltered for those that encode cell-surface proteins because these can be targeted by immunotherapies such as antibody-drug conjugate (ADC) [13] or chimeric antigen receptor (CAR) T-cell [14, 15]. The final candidate, GPC2, was experimentally validated by demonstrating that a GPC2-targeted ADC safely and effectively delivered cytotoxic drugs to eliminate neuroblastoma cells in-vitro and in-vivo. Similarly, another study [16] compared tissue-level gene expression from 12 pediatric cancers to normal tissue samples. Through individual identification of highly expressed membrane proteins in the cancerous tissues, targetable tumor antigens were identified. Novel candidates as well as targets that are currently being tested in clinical trials were identified, including GPC2. Other studies integrated transcriptomics and proteomics profiling to identify and rank differentially expressed cell surface proteins in prostate cancer [17] and multiple myeloma [18] cell lines. To make such approaches accessible to users without computational background, we developed TargetRanger [3], a bioinformatics webserver application that identified highly expressed genes from a tumor sample by comparing the expression levels in the tumor to the expression of the same genes in GTEx [12], ARCHS4 [19], and the Tabula sapiens [20] atlases.\n",
    "\n",
    "Here, we developed a workflow called Multiomics2Targets by applying X2K and TargetRanger to the transcriptomics, proteomics, and phosphoproteomics data collected from cohorts of cancer patients. The workflow first performs pathway analysis of the tumor samples using the X2K algorithm. The workflow accepts gene expression profiles from a collection of tumors, performs clustering analysis on the expression data matrix, and then identifies cell signaling pathways that are most likely activated and inhibited in each cluster. Since the X2K pipeline produces a ranked list of protein kinases, these ranked kinases are compared to ranked kinases inferred from applying the KEA3 tool [7] directly to the phosphoproteomics data collected from the same cluster of patients. The kinases that are identified by both tools are likely the most relevant to target.\n",
    "Then cell surface targets that are highly expressed in the tumor compared to the expression of these targets in normal cell types and tissues from different atlases are identified with TargetRanger [3]. The entire Multiomics2Targets workflow is illustrated in the diagram below (Fig.  1).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "\n",
    "{% if rna_file.value != '' and prot_file.value != '' and phospho_file.value != '' %}\n",
    "<img src='https://minio.dev.maayanlab.cloud/x2k-tr/X2KTR_workflows.png' alt=\"Multiomics2Targets Workflow Diagram\">\n",
    "{% elif rna_file.value != '' and prot_file.value != '' %}\n",
    "<img src='https://minio.dev.maayanlab.cloud/x2k-tr/X2KTR_workflows-rna-prot.png' alt=\"Multiomics2Targets Workflow Diagram\">\n",
    "{% elif rna_file.value != '' and phospho_file.value != '' %}\n",
    "<img src='https://minio.dev.maayanlab.cloud/x2k-tr/X2KTR_workflows-rna-phospho.png' alt=\"Multiomics2Targets Workflow Diagram\">\n",
    "{% elif phospho_file.value != '' and prot_file.value != '' %}\n",
    "<img src='https://minio.dev.maayanlab.cloud/x2k-tr/X2KTR_workflows-prot-phospho.png' alt=\"Multiomics2Targets Workflow Diagram\">\n",
    "{% elif phospho_file.value != '' %}\n",
    "<img src='https://minio.dev.maayanlab.cloud/x2k-tr/X2KTR_workflows-phospho.png' alt=\"Multiomics2Targets Workflow Diagram\">\n",
    "{% elif prot_file.value != '' %}\n",
    "<img src='https://minio.dev.maayanlab.cloud/x2k-tr/X2KTR_workflows-prot.png' alt=\"Multiomics2Targets Workflow Diagram\">\n",
    "{% else %}\n",
    "<img src='https://minio.dev.maayanlab.cloud/x2k-tr/X2KTR_workflows-rna.png' alt=\"Multiomics2Targets Workflow Diagram\">\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if rna_file.value != '' and prot_file.value != '' and phospho_file.value != '' %}\n",
    "display(Markdown(\"__Fig. 1__ Workflow of the Multiomics2Targets pipeline. The pipeline utilizes transcriptomics, phosphoproteomics, and proteomics data matrices from tumor samples to identify driver kinases and potential cell-surface targets.\"))\n",
    "{% elif rna_file.value != '' and prot_file.value != '' %}\n",
    "display(Markdown(\"__Fig. 1__ Workflow of the Multiomics2Targets pipeline. The pipeline utilizes transcriptomics and proteomics data matrices from tumor samples to identify driver kinases and potential cell-surface targets.\"))\n",
    "{% elif rna_file.value != '' and phospho_file.value != '' %}\n",
    "display(Markdown(\"__Fig. 1__ Workflow of the Multiomics2Targets pipeline. The pipeline utilizes transcriptomics and phosphoproteomics data matrices from tumor samples to identify driver kinases and potential cell-surface targets.\"))\n",
    "{% elif phospho_file.value != '' and prot_file != '' %}\n",
    "display(Markdown(\"__Fig. 1__ Workflow of the Multiomics2Targets pipeline. The pipeline utilizes proteomics and phosphoproteomics data matrices from tumor samples to identify driver kinase that could serve as potential targets.\"))\n",
    "{% elif phospho_file.value != '' %}\n",
    "display(Markdown(\"__Fig. 1__ Workflow of the Multiomics2Targets pipeline. The pipeline utilizes phosphoproteomics data matrices from tumor samples to identify driver kinases that could serve as potential targets.\"))\n",
    "{% elif prot_file.value != '' %}\n",
    "display(Markdown(\"__Fig. 1__ Workflow of the Multiomics2Targets pipeline. The pipeline utilizes proteomics data matrices from tumor samples to identify highly variable proteins in tumor samples.\"))\n",
    "{% else %}\n",
    "display(Markdown(\"__Fig. 1__ Workflow of the Multiomics2Targets pipeline. The pipeline utilizes transcriptomics data matrices from tumor samples to identify driver kinases and potential cell-surface targets.\"))\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "percent_rows_phospho = {{ percent_rows_phospho }}\n",
    "percent_rows_protein = {{ percent_rows_protein }}\n",
    "chea3_n = {{ chea3_n }}\n",
    "kea3_n = {{ kea3_n }}\n",
    "n_neighbors = {{ n_neighbors }}\n",
    "min_dist = {{ min_dist }}\n",
    "resolution = {{ resolution }}\n",
    "tumor_validate = {{ tumor_validate }}\n",
    "default_meta_cols = ['Sex', 'Tumor_Size_cm', 'Stage', 'Tobacco_smoking_history']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Methods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if rna_file.value != '' %}\n",
    "*Clustering Samples from the RNA-seq Matrix*\n",
    "\n",
    "Grouping patients together who share similar vectors of expression may lead to identification of targets specific to smaller subsets of patients. Clustering is performed on the 5000 most variable log2 normalized genes using the Leiden community algorithm [21] \n",
    "\n",
    "*Differentially Expressed Genes*\n",
    "\n",
    "Differentially expressed genes (DEGs) are computed limma voom [22] for each cluster versus all the other samples. Additionally, differentially expressed genes are submitted to Enrichr [23] for identification of biological functions from Gene Ontology (GO BP) [24], pathways from Wikipathways [25], and phenotypes from the Mouse Genome Informatics (MGI) Mammmalian Phenotype Onotology [26].\n",
    "\n",
    "*Consensus Cluster Labels*\n",
    "\n",
    "We utilize GPT-4 and the significantly enriched terms from Enrichr to compute 'consensus' labels for each identified cluster. \n",
    "The model is instructed to summarize the significant terms from the up- and down-regulated genes in five words or less. \n",
    "The system prompt is specified as follows: \"You are an assistant to a biologist who has performed enrichment analysis on a set of up-regulated and down-regulated genes for a certain cluster of patient samples. You must determine a consensus label for a given cluster based on the significantly enriched terms which relate to biological processes and phenotypes.\" The user prompt on each run is then formatted as follows:\n",
    "\"The most significantly enriched terms for the upregulated genes of cluster {cluster} are: {', '.join(up_enrichments)}. For the down genes the significantly enriched terms are: {', '.join(down_enrichments)}. Please provide a consensus label for this cluster with no other reasoning. The label should be at maximum 5 words in length.\"\n",
    "\n",
    "*Membrane Proteins Filter*\n",
    "\n",
    "A membrane/surface protein filter was developed from the intersection of two data sources:\n",
    "[COMPARTMENTS](https://compartments.jensenlab.org/About) [27] knowledge predictions for human genes where a filter was applied for \"Plasma membrane\" and \"Cell surface\" subcellular localization with a confidence score greater than or equal to 3, and the [Human Protein Atlas](https://www.proteinatlas.org/) [28] membrane proteins where a filter was applied for \"Evidence at protein level\" and removal of genes with \"Low tissue specificity.\"\n",
    "\n",
    "*Cell Surface Target Identification*\n",
    "\n",
    "We utilize the TargetRanger [3] to identify genes that are highly expressed in the target cells while lowly expressed across normal human cell types and tissues. TargetRanger [3] compares the inputted samples with processed RNA-seq from several atlases containing healthy tissue and cell type gene expression, namely, GTEx [12], ARCHS4 [19], and Tabula Sapiens [20].\n",
    "\n",
    "*Transcription Factor Enrichment Analysis with ChEA3*\n",
    "\n",
    "To assess enriched transcription factors (TFs) from individual samples, we normalized and Z-scored the uploaded RNA-seq data, and then submitted the top 500 upregulated and downregulated genes for each sample to ChEA3 [5] to retrieve ranked lists of TFs.\n",
    "\n",
    "*Compute TF intermediates using Protein-Protein Co-expression*\n",
    "\n",
    "To identify additional proteins that connect transcription factors to upstream kinases we utilize protein-protein co-expression from the Cancer Cell Line Encyclopedia (CCLE) [29] to augment the top-ranked enriched transcription factors.\n",
    "\n",
    "*Infering Kinases using KEA3 from TFs + Intermediates*\n",
    "\n",
    "To infer upstream kinases in the (Expression to Kinases) X2K [1,2] pipeline we submitted the co-expressed intermediates and enriched transcription factors to KEA3 [7].\n",
    "{% endif %}\n",
    "\n",
    "{% if phospho_file.value != '' %}\n",
    "*Phosphoproteomics Analysis*\n",
    "\n",
    "From the phosphoproteomic data, we can identify upregulated and downregulated kinases using kinase enrichment analysis from KEA3 [7], and then compare those kinases to the ones identified by the X2K analysis.\n",
    "{% endif %}\n",
    "\n",
    "{% if prot_file.value != '' %}\n",
    "*Proteomics Analysis*\n",
    "\n",
    "Using the uploaded proteomics data, we can further validate the kinases and cell-surface protein targets identified in the pipeline to ensure the proteins are also expressed at the protein level after Z-scoring the matrix. \n",
    "{% endif %}\n",
    "\n",
    "{% if rna_file.value != '' and tumor_validate != [] %}\n",
    "*Validation of Cancer Targets*\n",
    "To gain insight into target efficacy, the DepMap database [30] provides gene effect scores that quantify the effect of gene knock-down on cell viability. Gene effect scores are averaged from the cancer cell lines that match the specified tumor type.\n",
    "To gain insight into target novelty, we query PubMed for the number of literature co-occurrences that exist between the gene target and the specified cancer.  \n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if rna_file.value != '' %}\n",
    "### Clustering Samples from the RNA-seq Expression Matrix\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:\n",
    "    rna_df.index = rna_df.index.map(lambda g: g.split('.')[0])\n",
    "    rna_df.index = rna_df.index.map(lambda g: lookup(g) if lookup(g) else g)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:    \n",
    "    adata = sc.AnnData(log2_normalize(qnorm.quantile_normalize(rna_df)).T.values, dtype=\"float\")\n",
    "    adata.var['gene_names'] = rna_df.index.values\n",
    "    adata.obs['samples'] = rna_df.columns.values\n",
    "    # Sort genes by variance\n",
    "    adata.var['var_rank'] = (-np.var(adata.X, axis=0, dtype=\"float\")).argsort()\n",
    "    adata = adata[:, adata.var.var_rank < 5000]\n",
    "\n",
    "    # UMAP\n",
    "    sc.pp.pca(adata)\n",
    "    sc.pp.neighbors(adata, n_neighbors=n_neighbors)  # create neighborhood graph\n",
    "    sc.tl.umap(adata, min_dist=min_dist, alpha=0.1)  # embed umap based on neighborhood graph\n",
    "    sc.tl.leiden(adata, resolution=resolution)  # clustering\n",
    "    clusters = adata.obs['leiden'].unique()\n",
    "    cmap = plt.cm.get_cmap('tab20b', len(clusters))\n",
    "    for i, cluster in enumerate(sorted(clusters)):\n",
    "        mask = adata.obs['leiden'] == cluster\n",
    "        ax = plt.scatter(x=adata.obsm['X_umap'][mask, 0], y=adata.obsm['X_umap'][mask, 1], \n",
    "                    color=cmap(i), label=f'Cluster {cluster}', s=10)\n",
    "    plt.legend(title='Leiden clusters', bbox_to_anchor=(1.05, 1), loc='best')\n",
    "    ax.axes.set_xticks([])\n",
    "    ax.axes.set_yticks([])\n",
    "    plt.xlabel('UMAP-1')\n",
    "    plt.ylabel('UMAP-2')\n",
    "    plt.tight_layout()\n",
    "    plt.savefig(f'figures/{rna_filename}_umap_leiden.png')\n",
    "    plt.savefig(f'figures/{rna_filename}_umap_leiden.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:    \n",
    "    display(Markdown(f\"__Fig. {fig_counter}__ Leiden clusters determined from log2 RNA-seq expression for 5000 most variable genes.\"))\n",
    "    display(FileLink(f'figures/{rna_filename}_umap_leiden.png', result_html_prefix='Leiden clusters UMAP PNG: '))\n",
    "    display(FileLink(f'figures/{rna_filename}_umap_leiden.svg', result_html_prefix='Leiden clusters UMAP SVG: '))\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:\n",
    "    leiden_df = pd.DataFrame([adata.obs['leiden'], adata.obs['samples']]).T.set_index('samples')\n",
    "    clusters = list(leiden_df['leiden'].unique())\n",
    "    display(leiden_df.value_counts().to_frame())\n",
    "    display(Markdown(f\"__Table {table_counter}__ Cluster distribution of samples.\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:\n",
    "    display(Markdown(f\"Based on the normalized most variable genes in RNA-seq data matrix {len(clusters)} clusters were identified (Fig. {fig_counter}). The number of samples in each cluster is shown in Table {table_counter}.\"))\n",
    "    fig_counter += 1\n",
    "    table_counter += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if rna_file.value != '' %}\n",
    "### Computing Differentially Expressed Genes\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:\n",
    "    cluster_dges = {}\n",
    "    dge_table = {}\n",
    "    for c in sorted(clusters):\n",
    "        cluster_dges[c] = {}\n",
    "        cluster_samples = list(leiden_df[leiden_df['leiden'] == c].index.values)\n",
    "        other_samples = [s for s in rna_df.columns if s not in cluster_samples]\n",
    "\n",
    "        with suppress_output():\n",
    "            dge = limma_voom_differential_expression(\n",
    "                rna_df[other_samples], rna_df[cluster_samples],\n",
    "                voom_design=True,\n",
    "            )\n",
    "        dge_filename = f'results/{rna_filename}_cluster_{c}_vs_rest_dge_limma.tsv'\n",
    "        dge.sort_values(by=\"adj.P.Val\", ascending=True).to_csv(dge_filename, sep='\\t')\n",
    "        sig_dge = dge[dge['adj.P.Val'] < 0.01]\n",
    "        sig_up_gs = list(sig_dge[sig_dge['logFC'] > 0].index.values)\n",
    "        sig_dn_gs = list(sig_dge[sig_dge['logFC'] < 0].index.values)\n",
    "        if len(sig_up_gs) > 2000:\n",
    "            sig_up_gs = sig_up_gs[:2000]\n",
    "        elif len(sig_up_gs) < 50:\n",
    "            sig_dge = dge[dge['adj.P.Val'] < 0.05]\n",
    "            sig_up_gs = list(sig_dge[sig_dge['logFC'] > 0].index.values)\n",
    "        if len(sig_dn_gs) > 2000:\n",
    "            sig_dn_gs = sig_dn_gs[:2000]\n",
    "        elif len(sig_dn_gs) < 50:\n",
    "            sig_dge = dge[dge['adj.P.Val'] < 0.05]\n",
    "            sig_dn_gs = list(sig_dge[sig_dge['logFC'] > 0].index.values)\n",
    "        cluster_dges[c]['up'] = sig_up_gs\n",
    "        cluster_dges[c]['down'] = sig_dn_gs\n",
    "        dge_table[c] = {}\n",
    "        dge_table[c]['up'] = len(cluster_dges[c]['up'])\n",
    "        dge_table[c]['down'] = len(cluster_dges[c]['down'])\n",
    "        display(FileLink(dge_filename, result_html_prefix=f'Differential expression for cluster {c}: '))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:\n",
    "    enrichr_lists = {}\n",
    "    for c in sorted(clusters):\n",
    "        try:\n",
    "            # filter for only protein coding genes to match Enrichr background.\n",
    "            up_protein_coding = [g for g in cluster_dges[c]['up'] if lookup(g)]\n",
    "            dn_protein_coding = [g for g in cluster_dges[c]['down'] if lookup(g)]\n",
    "            up = enrichr_link_from_genes(up_protein_coding, f'Upregulated genes in cluster {c}')\n",
    "            dn = enrichr_link_from_genes(dn_protein_coding, f'Downregulated genes in cluster {c}')\n",
    "            dge_table[c]['Enrichr Up'] = up['link']\n",
    "            dge_table[c]['Enrichr Down'] = dn['link']\n",
    "            enrichr_lists[c] = (up, dn)\n",
    "        except Exception as e:\n",
    "            display(f'Error processing cluster {c}:', e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:\n",
    "    dge_table_df = pd.DataFrame(dge_table).T\n",
    "    dge_table_df.index.name = 'Cluster'\n",
    "    dge_table_df = dge_table_df.style.format({'Enrichr Up': lambda url: f'<a href=\"{url}\" rel=\"noopener noreferrer\" target=\"_blank\">Enrichr - up genes</a>',\n",
    "                                              'Enrichr Down': lambda url: f'<a href=\"{url}\" rel=\"noopener noreferrer\" target=\"_blank\">Enrichr - down genes</a>'})\n",
    "    display(dge_table_df)\n",
    "    display(Markdown(f\"__Table {table_counter}__ Number of upregulated and downregulated genes in each cluster.\"))\n",
    "    table_counter += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:\n",
    "    enrichr_labels = {}\n",
    "    letter_counter = 0\n",
    "    for cluster in enrichr_lists:\n",
    "        enrichr_labels[cluster] = {}\n",
    "        up, dn = enrichr_lists[cluster]\n",
    "        display(Markdown(f'### Cluster {cluster}'))\n",
    "        display(Markdown(f'#### Upregulated genes'))\n",
    "        try:\n",
    "            res_list = enrich_libraries(up['userListId'])\n",
    "            enrichr_labels[cluster]['up'] = res_list\n",
    "            fig = enrichr_figure(res_list)\n",
    "            fig.savefig(f'figures/enrichr_up_cluster_{cluster}.png', bbox_inches='tight')\n",
    "            fig.savefig(f'figures/enrichr_up_cluster_{cluster}.svg', bbox_inches='tight')\n",
    "            display(Markdown(f\"__Fig. {fig_counter}{ascii_uppercase[letter_counter]}__ Enrichr results for up-regulated genes in cluster {cluster} from GO BP, Wikipathways, and MGI. \\* indicates adjusted p-value < 0.01.\"))\n",
    "            display(FileLink(f'figures/enrichr_up_cluster_{cluster}.png', result_html_prefix='Enrichr figure for upregulated genes in cluster PNG: '))\n",
    "            display(FileLink(f'figures/enrichr_up_cluster_{cluster}.svg', result_html_prefix='Enrichr figure for upregulated genes in cluster SVG: '))\n",
    "            display(Markdown(f'#### Downregulated genes'))\n",
    "            letter_counter += 1\n",
    "        except Exception as e:\n",
    "            print(f'Error processing cluster {cluster} upregulated genes:', e)\n",
    "        try:\n",
    "            res_list = enrich_libraries(dn['userListId'])\n",
    "            enrichr_labels[cluster]['down'] = res_list\n",
    "            fig = enrichr_figure(res_list)\n",
    "            fig.savefig(f'figures/enrichr_dn_cluster_{cluster}.png', bbox_inches='tight')\n",
    "            fig.savefig(f'figures/enrichr_dn_cluster_{cluster}.svg', bbox_inches='tight')\n",
    "            display(Markdown(f\"__Fig. {fig_counter}{ascii_uppercase[letter_counter]}__ Enrichr results for down-regulated genes in cluster {cluster} from GO BP, Wikipathways, and MGI. \\* indicates adjusted p-value < 0.01.\"))\n",
    "            display(FileLink(f'figures/enrichr_dn_cluster_{cluster}.png', result_html_prefix='Enrichr figure for downregulated genes in cluster PNG: '))\n",
    "            display(FileLink(f'figures/enrichr_dn_cluster_{cluster}.svg', result_html_prefix='Enrichr figure for downregulated genes in cluster SVG: '))\n",
    "            letter_counter += 1\n",
    "        except Exception as e:\n",
    "            print(f'Error processing cluster {cluster} downregulated genes:', e)\n",
    "        \n",
    "fig_counter += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if rna_file.value != '' %}\n",
    "### Consensus Cluster Labels\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:\n",
    "    labels = label_clusters(enrichr_labels)\n",
    "    discussion_results['Leiden Clusters Labeled with GPT from Enrichr Pathways & Phenotypes'] = convert_to_string(labels)\n",
    "    adata.obs['label'] = adata.obs['leiden'].map(lambda c: labels.get(c, 'Unlabled') + f' (C{c})')\n",
    "    cmap = plt.cm.get_cmap('tab20b', len(clusters))\n",
    "    for i, cluster in enumerate(sorted(clusters)):\n",
    "        mask = adata.obs['leiden'] == cluster\n",
    "        ax = plt.scatter(x=adata.obsm['X_umap'][mask, 0], y=adata.obsm['X_umap'][mask, 1], \n",
    "                    color=cmap(i), label=f\"{labels.get(cluster, 'Unlabeled')} (C{cluster})\", s=10)\n",
    "    plt.legend(title='Leiden clusters', bbox_to_anchor=(1.05, 1), loc='best')\n",
    "    ax.axes.set_xticks([])\n",
    "    ax.axes.set_yticks([])\n",
    "    plt.xlabel('UMAP-1')\n",
    "    plt.ylabel('UMAP-2')\n",
    "    plt.savefig(f'figures/{rna_filename}_clusters_labeled.png', bbox_inches='tight')\n",
    "    plt.savefig(f'figures/{rna_filename}_clusters_labeled.svg', bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:\n",
    "    display(Markdown(f\"__Fig. {fig_counter}__ Consensus labels created using GPT-4 and significantly enriched terms for differentially expressed genes from leiden clusters.\"))\n",
    "    display(FileLink(f'figures/{rna_filename}_clusters_labeled.png', result_html_prefix='Clusters labeled UMAP PNG: '))\n",
    "    display(FileLink(f'figures/{rna_filename}_clusters_labeled.svg', result_html_prefix='Clusters labeled UMAP SVG: '))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:\n",
    "    display(Markdown(f\"Consensus labels were created for each cluster based on the most significantly enriched terms from the up- and down-regulated genes (Fig. {fig_counter -1}{ascii_uppercase[0]}-{ascii_uppercase[letter_counter]}) and are shown in Fig. {fig_counter}.\"))\n",
    "    fig_counter += 1\n",
    "    letter_counter = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:\n",
    "    display(Markdown(describe_clusters(clean_enrichr_lables(enrichr_labels))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:\n",
    "    display(Markdown(f\"*\\*This section was generated with GPT-4 and should be interpreted with caution*\\*\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if rna_file.value != '' %}\n",
    "### Metadata Analysis\n",
    "Using the uploaded metadata file, we can observe the relative occurrences of different metadata categories in the dataset per cluster.\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file and has_meta_file:\n",
    "    metadata_cluster_values = {}\n",
    "    categorical_labels = set()\n",
    "    if 'data_type' in meta_df.index:\n",
    "        meta_df = meta_df.drop(index='data_type')\n",
    "    for c in sorted(clusters):\n",
    "        metadata_cluster_values[c] = []\n",
    "        cluster_samples = list(set(leiden_df[leiden_df['leiden'] == c].index.values).intersection(meta_df.index))\n",
    "        meta_cluster = meta_df.loc[cluster_samples]\n",
    "        numeric_data = []\n",
    "        categorical_data = []\n",
    "        for col in meta_cluster.columns:\n",
    "            try:\n",
    "                numeric_data.append([col, meta_cluster[col].astype(float).mean().round(2), meta_cluster[col].astype(float).std().round(2)])\n",
    "                metadata_cluster_values[c].append({'Attribute': col, 'Mean': meta_cluster[col].astype(float).mean(), 'Std': meta_cluster[col].astype(float).std()})\n",
    "            except:\n",
    "                categorical_labels.add(col)\n",
    "                categorical_data.append([col, meta_cluster[col].value_counts()])\n",
    "        numeric_data_df = pd.DataFrame(numeric_data, columns=['Attribute', 'Mean', 'Std'])\n",
    "        if not numeric_data_df.empty:\n",
    "            numeric_data= numeric_data_df.set_index('Attribute')\n",
    "            display(numeric_data)\n",
    "            display(Markdown(f\"__Table {table_counter}__ Numeric metadata for cluster {c}\"))\n",
    "            table_counter += 1\n",
    "        else:\n",
    "            display(Markdown(f\"No metadata was available for cluster {c} or an error occurred.\"))\n",
    "        data = []\n",
    "        for name, value_count in categorical_data:\n",
    "            total_count = value_count.sum()\n",
    "            for index, count in value_count.items():\n",
    "                percentage = (count / total_count) * 100\n",
    "                data.append({'Category': name, 'Value': index, 'Count': count, 'Percentage': percentage.round(1)})\n",
    "        result_df = pd.DataFrame(data)\n",
    "        if not result_df.empty:\n",
    "            grouped_df = result_df.set_index(['Category', 'Value'])\n",
    "            display(grouped_df)\n",
    "            display(Markdown(f\"__Table {table_counter}__ Categorical metadata for cluster {c}\"))\n",
    "            table_counter += 1\n",
    "        else:\n",
    "            display(Markdown(f\"No metadata was available for cluster {c} or an error occurred.\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file and has_meta_file:\n",
    "    try:\n",
    "        results = []\n",
    "        n_pop = len(meta_df)\n",
    "        for cluster_key, cluster_data in metadata_cluster_values.items():\n",
    "            n_clus = len(leiden_df[leiden_df['leiden'] == cluster_key])\n",
    "            for attribute in cluster_data:\n",
    "                attr_name = attribute['Attribute']\n",
    "                attr_mean = attribute['Mean']\n",
    "                attr_std = attribute['Std']\n",
    "                population_mean = meta_df[attr_name].astype(float).mean()\n",
    "                population_std = meta_df[attr_name].astype(float).std()\n",
    "            \n",
    "                t_statistic, p_value = ttest_ind_from_stats(attr_mean, attr_std, n_clus, population_mean, population_std, n_pop, equal_var=False)\n",
    "                if np.isnan(p_value):\n",
    "                    p_value = 1\n",
    "                    t_statistic = 0\n",
    "                result = {\n",
    "                    'Cluster': cluster_key,\n",
    "                    'Metadata Attribute': attr_name,\n",
    "                    'Cluster Mean': attr_mean,\n",
    "                    'All Samples Mean': population_mean,\n",
    "                    'T-Statistic': t_statistic,\n",
    "                    'P-Value': p_value\n",
    "                }\n",
    "                results.append(result)\n",
    "\n",
    "        # Create a DataFrame from the results list\n",
    "        df_results = pd.DataFrame(results)\n",
    "\n",
    "        _, pvalues_adj, _, _ = multipletests(df_results['P-Value'].values, method='fdr_bh')\n",
    "        df_results['Adj. P-Value'] = pvalues_adj\n",
    "        df_meta_sorted = df_results.set_index('Cluster').sort_values(['Cluster', 'Adj. P-Value'])\n",
    "        df_meta_sorted_sig = df_meta_sorted[df_meta_sorted['Adj. P-Value'] < 0.05]\n",
    "        if not df_meta_sorted_sig.empty:\n",
    "            df_meta_sorted_sig.to_csv('results/sig_metadata_attrs_cluster.tsv', sep='\\t')\n",
    "            for cluster in metadata_cluster_values:\n",
    "                if df_meta_sorted_sig[df_meta_sorted_sig.index == cluster].empty:\n",
    "                    display(Markdown(f\"No significant metadata attributes found for cluster {cluster}.\"))\n",
    "                    continue\n",
    "                display(df_meta_sorted_sig[df_meta_sorted_sig.index == cluster].head(10))\n",
    "                display(Markdown(f\"__Table {table_counter}__ Significant up and down attributes for numerical metadata for cluster {cluster}.\"))\n",
    "                table_counter += 1\n",
    "            display(FileLink('results/sig_metadata_attrs_cluster.tsv', result_html_prefix='Significant up and down attributes for numerical metadata for each cluster: '))\n",
    "        else:\n",
    "            display(df_meta_sorted)\n",
    "            display(Markdown(\"No significant metadata attributes found.\"))\n",
    "    except Exception as e:\n",
    "        print('Error computing significance for metadata attributes in each cluster:', e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if rna_file.value != '' and membrane_screener_.raw_value %}\n",
    "proteins = pd.read_csv('https://appyters.maayanlab.cloud/storage/Tumor_Gene_Target_Screener/surfaceome.csv')\n",
    "membrane_proteins = list(proteins['genename'].map(lookup).dropna().values)\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if rna_file.value != '' %}\n",
    "### Cell Surface Target Identification\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "\n",
    "if has_rna_file:    \n",
    "    targets = {}\n",
    "    bgs = {'GTEx': 'gtex-gene-stats.tsv', 'ARCHS4': 'archs4-gene-stats.tsv', 'TS': 'ts_10x_cell-ontology-class_donors_tissue-labels_v1.tsv'}\n",
    "    def find_targets(rna_df, bg, targets):\n",
    "        df_bg_stats = pd.read_csv(f\"https://appyters.maayanlab.cloud/storage/Tumor_Gene_Target_Screener/{bgs[bg]}\", sep='\\t', index_col=[0,1])\n",
    "        df_bg_genes = df_bg_stats.unstack().index.map(lambda idx: lookup(idx.partition('.')[0]))\n",
    "        df_bg_stats = df_bg_stats.unstack().groupby(df_bg_genes, observed=True).median().stack()\n",
    "        df_bg_expr = df_bg_stats.loc[(slice(None), ['25%', '50%', '75%']), :].unstack()\n",
    "        common_index = list(set(rna_df.index) & set(df_bg_expr.index))\n",
    "        expr_df = rna_df.loc[common_index, :]\n",
    "        expr_df.reset_index(inplace=True)\n",
    "        expr_df.drop_duplicates(subset='gene_id', inplace=True)\n",
    "        expr_df.set_index('gene_id', inplace=True, drop=True)\n",
    "        {% if normalize_to_bg.raw_value %}\n",
    "        target_distribution = df_bg_expr.loc[common_index, :].median(axis=1)\n",
    "        df_expr_norm = qnorm.quantile_normalize(expr_df.loc[common_index, :], target=target_distribution)\n",
    "        df_bg_expr_norm = qnorm.quantile_normalize(df_bg_expr.loc[common_index, :], target=target_distribution)\n",
    "        {% else %}\n",
    "        df_expr_norm = expr_df.loc[common_index, :]\n",
    "        df_bg_expr_norm = df_bg_expr.loc[common_index, :]\n",
    "        {% endif %}\n",
    "        for cluster in tqdm(list(leiden_df['leiden'].unique())):\n",
    "            if cluster not in targets: targets[cluster] = {}\n",
    "            cluster_samples = list(leiden_df[leiden_df['leiden'] == cluster].index.values)\n",
    "            with suppress_output():\n",
    "                df_bg_expr_norm.columns = df_bg_expr_norm.columns.to_flat_index().map(lambda s: ', '.join(s))\n",
    "                dge = limma_voom_differential_expression(\n",
    "                    df_bg_expr_norm, df_expr_norm[cluster_samples],\n",
    "                    voom_design=True,\n",
    "                )\n",
    "                targets[cluster][bg] = dge[(dge['adj.P.Val'] < 0.01) & (dge['t'] > 0)].sort_values('t', ascending=False).index.values "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file: \n",
    "    print('Finding Targets using ARCHS4')\n",
    "    find_targets(rna_df, 'ARCHS4', targets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:    \n",
    "    print('Finding Targets using GTEx')\n",
    "    find_targets(rna_df, 'GTEx', targets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:      \n",
    "    print('Finding Targets using Tabula Sapiens')\n",
    "    find_targets(rna_df, 'TS', targets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_eval\n",
    "\n",
    "if has_rna_file:\n",
    "    YlGnBu = cm.get_cmap('YlGnBu_r', 8)\n",
    "    cmap = {\"None\":YlGnBu(0), \"ARCHS4\": YlGnBu(1), \"GTEx\":YlGnBu(2), \"TS\": YlGnBu(4), \"ARCHS4-GTEx\":YlGnBu(3),  \"ARCHS4-TS\": YlGnBu(5), \"GTEx-TS\": YlGnBu(6), \"All\": YlGnBu(7)}\n",
    "\n",
    "    top_targets_n = 100\n",
    "    target_list = []\n",
    "    for cluster in targets:\n",
    "        for bg in targets[cluster]:\n",
    "            {% if rna_file.value != '' and membrane_screener_.raw_value %}\n",
    "            target_list.extend(list(filter(lambda g: g in membrane_proteins, targets[cluster][bg][:top_targets_n])))\n",
    "            {% else %}\n",
    "            target_list.extend(list(targets[cluster][bg][:top_targets_n]))\n",
    "            {% endif %}\n",
    "    data1, data2, data3 = [], [], []\n",
    "    similarity = []\n",
    "    target_list = list(set(target_list))\n",
    "\n",
    "    for gene in target_list:\n",
    "        a = [1 if gene in targets[c]['ARCHS4'][:top_targets_n] else 0 for c in clusters]\n",
    "        g = [2 if gene in targets[c]['GTEx'][:top_targets_n] else 0 for c in clusters]\n",
    "        l = [4 if gene in targets[c]['TS'][:top_targets_n] else 0 for c in clusters]\n",
    "        data1.append(a)\n",
    "        data2.append(g)\n",
    "        data3.append(l)\n",
    "        similarity.append(np.dot(np.dot(np.array(a), np.array(g)), np.array(l)))\n",
    "\n",
    "    data = np.add(np.add(data1, data2), data3)\n",
    "\n",
    "    membrane_target_mat = pd.DataFrame(data)\n",
    "    membrane_target_mat.columns = [f\"Cluster {c}\" for c in clusters]\n",
    "    membrane_target_mat.index = target_list\n",
    "\n",
    "    membrane_target_mat['count'] = membrane_target_mat.sum(axis=1)\n",
    "    membrane_target_mat = membrane_target_mat[membrane_target_mat['count'] >= 7]\n",
    "    membrane_target_mat = membrane_target_mat.rename_axis('Membrane Target').sort_values(by = ['count', 'Membrane Target'], ascending = [False, True]).drop('count', axis=1)\n",
    "\n",
    "    h = membrane_target_mat.shape[0]\n",
    "\n",
    "    cluster_targets= {}\n",
    "    for col in membrane_target_mat.columns:\n",
    "        cluster_targets[col + ' Cell Surface Target'] = ','.join(list(membrane_target_mat[col][membrane_target_mat[col] >= 5].index.values))\n",
    "\n",
    "    g = sns.clustermap(membrane_target_mat, figsize=(4,0.3*h+2*(h<15)), cmap=YlGnBu, cbar_pos=None, dendrogram_ratio=0.1-(h<40)*0.01*(h-30), row_cluster=False, xticklabels=True, yticklabels=True)\n",
    "    g.ax_row_dendrogram.legend(handles=[Rectangle((0, 0), 0, 0, color=val, label=key) for key, val in cmap.items()],\n",
    "                                    title='Background', loc='upper right')\n",
    "\n",
    "    plt.show()\n",
    "\n",
    "    g.savefig(f'figures/{rna_filename}_membrane_target_matrix.png', dpi=300)\n",
    "    g.savefig(f'figures/{rna_filename}_membrane_target_matrix.svg', dpi=300)\n",
    "    membrane_target_mat.to_csv(f'results/{rna_filename}_membrane_target_mat.csv')\n",
    "    display(Markdown(f\"__Fig. {fig_counter}__ Top 100 consensus targets for each cluster with adjusted p-value < 0.01 compared to ARCHS4, GTEx, and Tabula Sapiens (TS) healthy backgrounds with limma voom.\"))\n",
    "    display(FileLink(f'results/{rna_filename}_membrane_target_mat.csv', result_html_prefix='Download Membrane Target Matrix: '))\n",
    "    display(FileLink(f'figures/{rna_filename}_membrane_target_matrix.png', result_html_prefix='Membrane Target Matrix PNG: '))\n",
    "    display(FileLink(f'figures/{rna_filename}_membrane_target_matrix.svg', result_html_prefix='Membrane Target Matrix SVG: '))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:\n",
    "    display(Markdown(f\"Cell surface targets were identified for each cluster and are ordered by the number of healthy backgrounds they were upregulated compared to (Fig. {fig_counter}).\"))\n",
    "    fig_counter += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:\n",
    "    display(Markdown(create_results_text(cluster_targets)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:\n",
    "    display(Markdown(f\"*\\*This section was generated with GPT-4 and should be interpreted with caution*\\*\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:\n",
    "    rna_df_log2_zscore = zscore(log2_normalize(qnorm.quantile_normalize(rna_df)), axis=1)\n",
    "    rna_df_log2_zscore"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if rna_file.value != '' %}\n",
    "### Transcription Factor Enrichment Analysis with ChEA3\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:    \n",
    "    chea_res = {}\n",
    "    for patient in tqdm(rna_df_log2_zscore.columns.values, leave=False, mininterval=30):\n",
    "        chea_res[patient] = {}\n",
    "        down_rows = list(filter(lambda g: lookup(g) ,rna_df_log2_zscore[patient].astype(float).sort_values(ascending=True).index.values))[:chea3_n]\n",
    "        up_rows = list(filter(lambda g: lookup(g) ,rna_df_log2_zscore[patient].astype(float).sort_values(ascending=False).index.values))[:chea3_n]\n",
    "\n",
    "        chea_down = get_chea3_results(down_rows, f'{patient}_down_genes')\n",
    "        chea_up = get_chea3_results(up_rows, f'{patient}_up_genes')\n",
    "        chea_res[patient]['down'] = chea_down\n",
    "        chea_res[patient]['up'] = chea_up"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if rna_file.value != '' %}\n",
    "#### Full ChEA Heatmap - Up Genes\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:\n",
    "    up_chea_comb = pd.concat([chea_res[p]['up'] for p in chea_res], axis=1)\n",
    "    up_chea_comb.columns = list(chea_res.keys())\n",
    "    try:\n",
    "        g = create_annotated_clustermap(up_chea_comb, meta_df, leiden_df, default_meta_cols)\n",
    "    except:\n",
    "        plt.clf()\n",
    "        g = create_clustermap(up_chea_comb.astype(float), cbar_label='Mean Rank', ylabel='Transcription Factors')\n",
    "    clustergrammer_res = clustergrammer_link(up_chea_comb, f'{rna_filename}_chea_up_heatmap')\n",
    "    if clustergrammer_res:\n",
    "        display(Markdown(f\"[View in Clustergrammer]({clustergrammer_res})\"))\n",
    "    g.savefig(f'figures/{rna_filename}_chea_up_heatmap.png', dpi=300)\n",
    "    g.savefig(f'figures/{rna_filename}_chea_up_heatmap.svg', dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:\n",
    "    display(Markdown(f\"__Fig. {fig_counter}{ascii_uppercase[letter_counter]}__ Integrated Mean Rank from ChEA3 from up-regulated genes for all samples and all transcription factors based on z-scored expression.\"))\n",
    "    letter_counter += 1\n",
    "    display(FileLink(f'figures/{rna_filename}_chea_up_heatmap.png', result_html_prefix='ChEA3 Up Heatmap PNG: '))\n",
    "    display(FileLink(f'figures/{rna_filename}_chea_up_heatmap.svg', result_html_prefix='ChEA3 Up Heatmap SVG: '))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if rna_file.value != '' %}\n",
    "### Top 20 Variable TFs - ChEA Heatmap (Up Genes)\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file: \n",
    "    top_20_var_tfs = up_chea_comb.var(axis=1).sort_values(ascending=False).index[:20]\n",
    "    try:\n",
    "        g = create_annotated_clustermap(up_chea_comb.loc[top_20_var_tfs].astype(float), meta_df, leiden_df, default_meta_cols, cbar_label='Mean Rank', ylabel='Top 20 TFs by Variance', yticklabels=True)\n",
    "    except:\n",
    "        plt.clf()\n",
    "        g = create_clustermap(up_chea_comb.loc[top_20_var_tfs].astype(float), ylabel='Top 20 TFs by Variance', cbar_label='Mean Rank', yticklabels=True)\n",
    "    clustergrammer_res = clustergrammer_link(up_chea_comb.loc[top_20_var_tfs], f'{rna_filename}_chea_up_heatmap_top_var_tfs')\n",
    "    if clustergrammer_res:\n",
    "        display(Markdown(f\"[View in Clustergrammer]({clustergrammer_res})\"))\n",
    "    g.savefig(f'figures/{rna_filename}_top_var_tfs_up_heatmap.png', dpi=300)\n",
    "    g.savefig(f'figures/{rna_filename}_top_var_tfs_up_heatmap.svg', dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:    \n",
    "    display(Markdown(f\"__Fig. {fig_counter}{ascii_uppercase[letter_counter]}__ Integrated Mean Rank from ChEA3 from up-regulated genes for all samples and the top 20 most variable transcription factors based on z-scored expression.\"))\n",
    "    letter_counter += 1\n",
    "    display(FileLink(f'figures/{rna_filename}_top_var_tfs_up_heatmap.png', result_html_prefix='Most Variable TFs Heatmap (up genes) PNG: '))\n",
    "    display(FileLink(f'figures/{rna_filename}_top_var_tfs_up_heatmap.svg', result_html_prefix='Most Variable TFs Heatmap (up genes) SVG: '))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if rna_file.value != '' %}\n",
    "#### Full ChEA Heatmap - Down Genes\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:    \n",
    "    down_chea_comb = pd.concat([chea_res[p]['down'] for p in chea_res], axis=1)\n",
    "    down_chea_comb.columns = list(chea_res.keys())\n",
    "    try:\n",
    "        g = create_annotated_clustermap(down_chea_comb.astype(float), meta_df, leiden_df, default_meta_cols)\n",
    "    except:\n",
    "        plt.clf()\n",
    "        g = create_clustermap(down_chea_comb.astype(float), ylabel='Transcription Factors', cbar_label='Mean Rank')\n",
    "    clustergrammer_res = clustergrammer_link(down_chea_comb, f'{rna_filename}_chea_dn_heatmap')\n",
    "    if clustergrammer_res:\n",
    "        display(Markdown(f\"[View in Clustergrammer]({clustergrammer_res})\"))\n",
    "    g.savefig(f'figures/{rna_filename}_chea_dn_heatmap.png', dpi=300)\n",
    "    g.savefig(f'figures/{rna_filename}_chea_dn_heatmap.svg', dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:    \n",
    "    display(Markdown(f\"__Fig. {fig_counter}{ascii_uppercase[letter_counter]}__ Integrated Mean Rank from ChEA3 from down-regulated genes for all samples and all transcription factors based on z-scored expression.\"))\n",
    "    letter_counter += 1\n",
    "    display(FileLink(f'figures/{rna_filename}_chea_dn_heatmap.png', result_html_prefix='ChEA3 Down Heatmap PNG: '))\n",
    "    display(FileLink(f'figures/{rna_filename}_chea_dn_heatmap.svg', result_html_prefix='ChEA3 Down Heatmap SVG: ')) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if rna_file.value != '' %}\n",
    "### Top 20 Variable TFs - ChEA Heatmap (Down Genes)\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:    \n",
    "    top_20_var_tfs = down_chea_comb.var(axis=1).sort_values(ascending=False).index[:20]\n",
    "    try:\n",
    "        g = create_annotated_clustermap(down_chea_comb.loc[top_20_var_tfs].astype(float), meta_df, leiden_df, default_meta_cols, cbar_label='Mean Rank', ylabel='Top 20 TFs by Variance', yticklabels=True)\n",
    "    except:\n",
    "        plt.clf()\n",
    "        g = create_clustermap(down_chea_comb.loc[top_20_var_tfs].astype(float), cbar_label='Mean Rank', ylabel='Top 20 TFs by Variance', yticklabels=True)\n",
    "    clustergrammer_res = clustergrammer_link(down_chea_comb.loc[top_20_var_tfs].astype(float), f'{rna_filename}_chea_dn_heatmap_top_var_tfs')\n",
    "    if clustergrammer_res:\n",
    "        display(Markdown(f\"[View in Clustergrammer]({clustergrammer_res})\"))\n",
    "    g.savefig(f'figures/{rna_filename}_top_var_tfs_dn_heatmap.png', dpi=300)\n",
    "    g.savefig(f'figures/{rna_filename}_top_var_tfs_dn_heatmap.svg', dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:    \n",
    "    display(Markdown(f\"__Fig. {fig_counter}{ascii_uppercase[letter_counter]}__ Integrated Mean Rank from ChEA3 from down-regulated genes for all samples and the top 20 most variable transcription factors based on z-scored expression.\"))\n",
    "    display(FileLink(f'figures/{rna_filename}_top_var_tfs_dn_heatmap.png', result_html_prefix='Most Variable TFs Heatmap (down genes) PNG: '))\n",
    "    display(FileLink(f'figures/{rna_filename}_top_var_tfs_dn_heatmap.svg', result_html_prefix='Most Variable TFs Heatmap (down genes) SVG: '))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:\n",
    "    display(Markdown(f\"ChEA3 vectors for Z-scored gene expression for each sample is visualized for both the up- and down-regulated genes as well as for the top 20 TFs by variance (Fig. {fig_counter}{ascii_uppercase[0]}-{fig_counter}{ascii_uppercase[letter_counter]}).\"))\n",
    "    letter_counter = 0\n",
    "    fig_counter += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if rna_file.value != '' %}\n",
    "### Computing TF intermediates using Protein-Protein Co-Expression\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:    \n",
    "    ids = list(chea_res.keys())\n",
    "    n_TFs = 20\n",
    "    def get_protein_coexpression():\n",
    "        protein_coexp = pd.read_csv('https://minio.dev.maayanlab.cloud/x2k-tr/ccleproteomics_normalized_gene_correlation_matrix.txt.gz', sep='\\t', compression='gzip', index_col=0)\n",
    "        up_TFs = {}\n",
    "        up_intermediates = {}\n",
    "        dn_TFs = {}\n",
    "        dn_intermediates = {}\n",
    "\n",
    "        for id in tqdm(ids, leave=False, mininterval=30, desc=\"Augmenting TFs...\"):\n",
    "            up_TFs[id] = chea_res[id]['up'].astype(float).sort_values(by='Score', ascending=True).index[:n_TFs].to_list()\n",
    "            up_intermediates[id] = augment_proteomics(gene_list=up_TFs[id], protein_coexp=protein_coexp)\n",
    "            dn_TFs[id] = chea_res[id]['down'].astype(float).sort_values(by='Score', ascending=True).index[:n_TFs].to_list()\n",
    "            dn_intermediates[id] = augment_proteomics(gene_list=dn_TFs[id], protein_coexp=protein_coexp)\n",
    "        \n",
    "        up_TFs = pd.DataFrame(up_TFs)\n",
    "        up_intermediates = pd.DataFrame(up_intermediates)\n",
    "        up_TFs.to_csv(f\"results/up_TFs.csv\", index=False)\n",
    "        up_intermediates.to_csv(f\"results/up_intermediates.csv\", index=False)\n",
    "\n",
    "        dn_TFs = pd.DataFrame(dn_TFs)\n",
    "        dn_intermediates = pd.DataFrame(dn_intermediates)\n",
    "        dn_TFs.to_csv(f\"results/dn_TFs.csv\", index=False)\n",
    "        dn_intermediates.to_csv(f\"results/dn_intermediates.csv\", index=False)\n",
    "        return up_TFs, up_intermediates, dn_TFs, dn_intermediates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:\n",
    "    up_TFs, up_intermediates, dn_TFs, dn_intermediates = get_protein_coexpression()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if rna_file.value != '' %}\n",
    "### Inferring Kinases using KEA3 from TFs + Intermediates\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:    \n",
    "    # Compute normalized kinase enrichment scores\n",
    "    up_kinases = {}\n",
    "    dn_kinases = {}\n",
    "    x2k_up = {}\n",
    "    x2k_dn = {}\n",
    "    \n",
    "    for id in tqdm(ids, leave=False, mininterval=30, desc=\"Computing RNA-seq derived KEA scores...\"):\n",
    "        up_geneset = list(set(list(up_TFs[id]) + list(filter(lambda x: x is not None, up_intermediates[id]))))\n",
    "        dn_geneset = list(set(list(dn_TFs[id]) + list(filter(lambda x: x is not None, dn_intermediates[id]))))\n",
    "        up_results = get_kea3_results(up_geneset)\n",
    "        dn_results = get_kea3_results(dn_geneset)\n",
    "        up_kinases[id] = up_results.astype(float).sort_values(by=\"Score\", ascending=True).index.to_list()\n",
    "        dn_kinases[id] = dn_results.astype(float).sort_values(by=\"Score\", ascending=True).index.to_list()\n",
    "        if len(x2k_up) == 0:\n",
    "            KINASES = up_results.Score.index.to_list()\n",
    "        x2k_up[id] = up_results.Score.loc[KINASES]\n",
    "        x2k_dn[id] = dn_results.Score.loc[KINASES]\n",
    "\n",
    "\n",
    "    pd.DataFrame(up_kinases).to_csv(f\"results/up_kinases.csv\", index=False)\n",
    "    pd.DataFrame(dn_kinases).to_csv(f\"results/dn_kinases.csv\", index=False)\n",
    "    x2k_up = pd.DataFrame(x2k_up, index=KINASES)\n",
    "    x2k_dn = pd.DataFrame(x2k_dn, index=KINASES)\n",
    "    x2k_up.to_csv(f\"results/up_x2k_meanRank.csv\")\n",
    "    x2k_dn.to_csv(f\"results/dn_x2k_meanRank.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if rna_file.value != '' %}\n",
    "#### X2K Heatmap - Up Genes\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file: \n",
    "    try:\n",
    "        g = create_annotated_clustermap(x2k_up.astype(float), meta_df, leiden_df, default_meta_cols, cbar_label='Mean Rank', ylabel='Kinases', xlabel='Samples')\n",
    "    except:\n",
    "        plt.clf()\n",
    "        g = create_clustermap(x2k_up.astype(float), cbar_label='Mean Rank', ylabel='Kinases', xlabel='Samples')\n",
    "    clustergrammer_res = clustergrammer_link(x2k_up.astype(float), f'{rna_filename}_x2k_up')\n",
    "    if clustergrammer_res:\n",
    "        display(Markdown(f\"[View in Clustergrammer]({clustergrammer_res})\"))\n",
    "    g.savefig(f'figures/{rna_filename}_x2k_up_heatmap.png', dpi=300)\n",
    "    g.savefig(f'figures/{rna_filename}_x2k_up_heatmap.svg', dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:    \n",
    "    display(Markdown(f\"__Fig. {fig_counter}{ascii_uppercase[letter_counter]}__ Results of X2K (Expression2Kinases) pipeline for up-regulated genes visualized as hierarchically clustered heatmap for all kinases.\"))\n",
    "    letter_counter += 1\n",
    "    display(FileLink(f'figures/{rna_filename}_x2k_up_heatmap.png', result_html_prefix='X2K Up Heatmap PNG: '))\n",
    "    display(FileLink(f'figures/{rna_filename}_x2k_up_heatmap.svg', result_html_prefix='X2K Up Heatmap SVG: '))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if rna_file.value != '' %}\n",
    "#### X2K Heatmap Top 20 Variable Kinases - Up Genes\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:    \n",
    "    top_20_var_x2k = x2k_up.var(axis=1).sort_values(ascending=False).index[:20]\n",
    "    try:\n",
    "        g = create_annotated_clustermap(x2k_up.loc[top_20_var_x2k].astype(float), meta_df, leiden_df, default_meta_cols, cbar_label='Mean Rank', xlabel='Samples', ylabel='Top 20 Kinases by Variance', yticklabels=True)\n",
    "    except:\n",
    "        plt.clf()\n",
    "        g = create_clustermap(x2k_up.loc[top_20_var_x2k].astype(float), cbar_label='Mean Rank', ylabel='Top 20 Kinases by Variance', xlabel='Samples', yticklabels=True)\n",
    "    clustergrammer_res = clustergrammer_link(x2k_up.loc[top_20_var_x2k], f'{rna_filename}_x2k_up_top20')\n",
    "    if clustergrammer_res:\n",
    "        display(Markdown(f\"[View in Clustergrammer]({clustergrammer_res})\"))\n",
    "    g.savefig(f'figures/{rna_filename}_x2k_up_top20_heatmap.png', dpi=300)\n",
    "    g.savefig(f'figures/{rna_filename}_x2k_up_top20_heatmap.svg', dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:\n",
    "    display(Markdown(f\"__Fig. {fig_counter}{ascii_uppercase[letter_counter]}.__ Results of X2K (Expression2Kinases) pipeline for up-regulated genes visualized as hierarchically clustered heatmap for 20 most variable kinases.\"))\n",
    "    letter_counter += 1\n",
    "    display(FileLink(f'figures/{rna_filename}_x2k_up_top20_heatmap.png', result_html_prefix='X2K Up Top 20 Kinases Heatmap PNG: '))\n",
    "    display(FileLink(f'figures/{rna_filename}_x2k_up_top20_heatmap.svg', result_html_prefix='X2K Up Top 20 Kinases Heatmap SVG: ')) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if rna_file.value != '' %}\n",
    "#### X2K Heatmap - Down Genes\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:   \n",
    "    try: \n",
    "        g = create_annotated_clustermap(x2k_dn.astype(float), meta_df, leiden_df, default_meta_cols, cbar_label='Mean Rank', ylabel='Kinases', xlabel='Samples')\n",
    "    except:\n",
    "        plt.clf()\n",
    "        g = create_clustermap(x2k_dn.astype(float), cbar_label='Mean Rank', ylabel='Kinases', xlabel='Samples')\n",
    "    clustergrammer_res = clustergrammer_link(x2k_dn.astype(float), f'{rna_filename}_x2k_dn')\n",
    "    if clustergrammer_res:\n",
    "        display(Markdown(f\"[View in Clustergrammer]({clustergrammer_res})\"))\n",
    "    g.savefig(f'figures/{rna_filename}_x2k_dn_heatmap.png', dpi=300)\n",
    "    g.savefig(f'figures/{rna_filename}_x2k_dn_heatmap.svg', dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:\n",
    "    display(Markdown(f\"__Fig. {fig_counter}{ascii_uppercase[letter_counter]}__ Results of X2K (Expression2Kinases) pipeline for down-regulated genes visualized as hierarchically clustered heatmap for all kinases.\"))\n",
    "    letter_counter += 1\n",
    "    display(FileLink(f'figures/{rna_filename}_x2k_dn_heatmap.png', result_html_prefix='X2K Down Heatmap PNG: '))\n",
    "    display(FileLink(f'figures/{rna_filename}_x2k_dn_heatmap.svg', result_html_prefix='X2K Down Heatmap SVG: '))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if rna_file.value != '' %}\n",
    "#### X2K Heatmap Top 20 Variable Kinases - Down Genes\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:    \n",
    "    top_20_var_x2k = x2k_dn.var(axis=1).sort_values(ascending=False).index[:20]\n",
    "    try:\n",
    "        g = create_annotated_clustermap(x2k_dn.loc[top_20_var_x2k].astype(float), meta_df, leiden_df, default_meta_cols, cbar_label='Mean Rank', ylabel='Top 20 Kinases by Variance', xlabel='Samples', yticklabels=True)\n",
    "    except:\n",
    "        plt.clf()\n",
    "        g = create_clustermap(x2k_dn.loc[top_20_var_x2k].astype(float), cbar_label='Mean Rank', ylabel='Top 20 Kinases by Variance', xlabel='Samples', yticklabels=True)\n",
    "    clustergrammer_res = clustergrammer_link(x2k_dn.loc[top_20_var_x2k].astype(float), f'{rna_filename}_x2k_dn_top20')\n",
    "    if clustergrammer_res:\n",
    "        display(Markdown(f\"[View in Clustergrammer]({clustergrammer_res})\"))\n",
    "    g.savefig(f'figures/{rna_filename}_x2k_dn_top20_heatmap.png', dpi=300)\n",
    "    g.savefig(f'figures/{rna_filename}_x2k_dn_top20_heatmap.svg', dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:\n",
    "    display(Markdown(f\"__Fig. {fig_counter}{ascii_uppercase[letter_counter]}__ Results of X2K (Expression2Kinases) pipeline for down-regulated genes visualized as hierarchically clustered heatmap for 20 most variable kinases.\"))\n",
    "    display(FileLink(f'figures/{rna_filename}_x2k_dn_top20_heatmap.png', result_html_prefix='X2K Down Top 20 Kinases Heatmap PNG: '))\n",
    "    display(FileLink(f'figures/{rna_filename}_x2k_dn_top20_heatmap.svg', result_html_prefix='X2K Down Top 20 Kinases Heatmap SVG: '))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:\n",
    "    display(Markdown(f\"X2K results for each sample were visualized as hierarchically clustered heatmaps for all kinases and the top 20 most variable kinases up- and down-regulated genes (Fig. {fig_counter}{ascii_uppercase[0]}-{ascii_uppercase[letter_counter]}).\"))\n",
    "    letter_counter = 0\n",
    "    fig_counter += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if rna_file.value != '' %}\n",
    "### Visualize Common X2K Pathways\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if rna_file.value != '' %}\n",
    "#### Pathways inferred from up-genes:\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:\n",
    "    pathway_results = {}\n",
    "    rank_threshold = 30 # cut-off of top-ranked kinases considered as enriched in a sample\n",
    "    n_edges = 30 # desired number of edges to show in the plot\n",
    "    dir = \"up\" # enrichment direction to visualize\n",
    "    kinase_col = \"#FFD700\" # kinase color\n",
    "    tf_col = \"#F08080\" # TF color\n",
    "\n",
    "    # Build network \n",
    "    edges_df = pd.DataFrame(columns=['target', 'source'])\n",
    "    nodes_df = pd.DataFrame(columns=['id', 'type', 'color'])\n",
    "\n",
    "    TFs = pd.read_csv(f\"results/{dir}_TFs.csv\")\n",
    "    intermediates = pd.read_csv(f\"results/{dir}_intermediates.csv\")\n",
    "    kinases = pd.read_csv(f\"results/{dir}_kinases.csv\")\n",
    "    samples = kinases.columns.to_list()\n",
    "\n",
    "    TRANSCRIPTION_FACTORS = TFs.to_numpy().flatten().tolist()\n",
    "    KINASES = kinases.to_numpy().flatten().tolist()\n",
    "    for cluster in list(sorted(leiden_df['leiden'].unique())):\n",
    "        cluster_samples = list(leiden_df[leiden_df['leiden'] == cluster].index.values)\n",
    "        for id in cluster_samples:\n",
    "            new = pd.DataFrame([[tf, p] for p in intermediates[id].dropna() for tf in TFs[id].dropna()], columns=['target', 'source'])\n",
    "            edges_df = pd.concat([edges_df, new])\n",
    "            new = pd.DataFrame([[p, kinase] for kinase in kinases[id].dropna()[:rank_threshold] for p in intermediates[id].dropna()], columns=['target', 'source'])\n",
    "            edges_df = pd.concat([edges_df, new])\n",
    "        \n",
    "        edges_df = edges_df.groupby(['target', 'source']).size().reset_index()\n",
    "        edges_df.rename(columns={0: \"count\"}, inplace=True)\n",
    "        edges_df['label'] = edges_df['count']\n",
    "        edges_df = edges_df.loc[edges_df.source != edges_df.target]\n",
    "\n",
    "\n",
    "        thresh = edges_df.sort_values(by='count', ascending=False).iloc[n_edges]['count']\n",
    "\n",
    "        if (thresh * 4) <= len(cluster_samples):\n",
    "            display(Markdown(f'No common pathways were identified for at least 25% of samples in cluster {cluster}.'))\n",
    "            continue\n",
    "        edges_show = edges_df.loc[edges_df['count'] >= thresh]\n",
    "        \n",
    "        nodes = pd.unique(edges_show[['target', 'source']].to_numpy().flatten().tolist())\n",
    "        nodes = nodes[~pd.isnull(nodes)]\n",
    "        nodes_df = pd.DataFrame({'id': nodes})\n",
    "        nodes_df['type'] = 'intermediate'\n",
    "        nodes_df['color'] = \"#DCDCDC\"\n",
    "        nodes_df.loc[nodes_df['id'].isin(TRANSCRIPTION_FACTORS), 'type'] = 'TF'\n",
    "        nodes_df.loc[nodes_df['id'].isin(KINASES), 'type'] = 'kinase'\n",
    "        nodes_df.loc[nodes_df['type'] == 'TF', 'color'] = tf_col\n",
    "        nodes_df.loc[nodes_df['type'] == 'kinase', 'color'] = kinase_col\n",
    "        nodes_df['label'] = nodes_df['id']\n",
    "\n",
    "        # Create a graph\n",
    "        G = nx.Graph()\n",
    "\n",
    "        # Add nodes\n",
    "        for idx, row in nodes_df.iterrows():\n",
    "            G.add_node(row['id'], group=row['type'], color=row['color'])\n",
    "\n",
    "        # Add edges\n",
    "        for idx, row in edges_show.iterrows():\n",
    "            G.add_edge(row['source'], row['target'])\n",
    "\n",
    "        # Manually set positions\n",
    "        pos = {}\n",
    "        group_positions = {'kinase': 0.8, 'intermediate': 0.5, 'TF': 0.2} \n",
    "        group_nodes = {group: [] for group in group_positions.keys()}\n",
    "        for node, data in G.nodes(data=True):\n",
    "            group_nodes[data['group']].append(node)\n",
    "        for group, nodes in group_nodes.items():\n",
    "            x_positions = np.linspace(0, 1, len(nodes))\n",
    "            for x, node in zip(x_positions, nodes):\n",
    "                pos[node] = (x, group_positions[group])\n",
    "\n",
    "        nodes_data =G.nodes(data=True)\n",
    "        clus_pathway =  {'kinases':[], 'intermediates': [], 'TFs': []}\n",
    "        for n in nodes_data:\n",
    "            if n[1]['group'] == 'kinase':\n",
    "                clus_pathway['kinases'].append(n[0])\n",
    "            elif n[1]['group'] == 'intermediate':\n",
    "                clus_pathway['intermediates'].append(n[0])\n",
    "            elif n[1]['group'] == 'TF':\n",
    "                clus_pathway['TFs'].append(n[0])\n",
    "        pathway_results[f\"Common X2K Pathways - Upregulated Genes - Cluster {cluster}\"] = convert_to_string(clus_pathway)\n",
    "    \n",
    "        node_colors = [G.nodes[n]['color'] for n in G.nodes()]\n",
    "        if len(G.nodes()) > 60:\n",
    "            display(Markdown(f'Too many nodes were identified for {cluster}.'))\n",
    "            continue\n",
    "\n",
    "        if len(G.nodes()) >= 50:\n",
    "            node_size = 600\n",
    "            font_size = 4\n",
    "        elif len(G.nodes()) > 40:\n",
    "            node_size = 800\n",
    "            font_size = 6\n",
    "        else:\n",
    "            node_size = 1500\n",
    "            font_size = 8\n",
    "        plt.figure(figsize=(12, 5))\n",
    "        nx.draw_networkx_nodes(G, pos, node_color=node_colors, node_size=node_size)\n",
    "        nx.draw_networkx_edges(G, pos, edgelist=edges_show.to_records(index=False), width=1)\n",
    "        nx.draw_networkx_labels(G, pos, font_size=font_size, font_family='sans-serif')\n",
    "        plt.savefig(f'figures/{dir}_x2k_pathways_C{cluster}.png', dpi=300)\n",
    "        plt.savefig(f'figures/{dir}_x2k_pathways_C{cluster}.svg', dpi=300)\n",
    "        nx.write_adjlist(G, f'results/{dir}_x2k_pathways_C{cluster}.adjlist')\n",
    "        plt.show()\n",
    "        display(Markdown(f\"__Fig. {fig_counter}{ascii_uppercase[letter_counter]}__ Visualization of pathways commonly identified from up-genes from cluster {cluster}.\"))\n",
    "        letter_counter += 1\n",
    "        display(FileLink(f'figures/{dir}_x2k_pathways_C{cluster}.png', result_html_prefix='Pathways Inferred from Up-Genes PNG: '))\n",
    "        display(FileLink(f'figures/{dir}_x2k_pathways_C{cluster}.svg', result_html_prefix='Pathways Inferred from Up-Genes SVG: '))\n",
    "        display(FileLink(f'results/{dir}_x2k_pathways_C{cluster}.adjlist', result_html_prefix='Download Adjacency List: '))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:\n",
    "    display(Markdown(f\"Common X2K pathways were inferred from up-regulated genes for each cluster and visualized as network graphs (Fig. {fig_counter}{ascii_uppercase[0]}-{ascii_uppercase[letter_counter - 1]}).\"))\n",
    "    fig_counter += 1\n",
    "    letter_counter = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if rna_file.value != '' %}\n",
    "#### Pathways inferred from down-genes:\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:\n",
    "    rank_threshold = 30 # cut-off of top-ranked kinases considered as enriched in a sample\n",
    "    n_edges = 30 # desired number of edges to show in the plot\n",
    "    dir = \"dn\" # enrichment direction to visualize\n",
    "    kinase_col = \"#FFD700\" # kinase color\n",
    "    tf_col = \"#F08080\" # TF color\n",
    "\n",
    "    # Build network \n",
    "    edges_df = pd.DataFrame(columns=['target', 'source'])\n",
    "    nodes_df = pd.DataFrame(columns=['id', 'type', 'color'])\n",
    "\n",
    "    TFs = pd.read_csv(f\"results/{dir}_TFs.csv\")\n",
    "    intermediates = pd.read_csv(f\"results/{dir}_intermediates.csv\")\n",
    "    kinases = pd.read_csv(f\"results/{dir}_kinases.csv\")\n",
    "    samples = kinases.columns.to_list()\n",
    "\n",
    "    TRANSCRIPTION_FACTORS = TFs.to_numpy().flatten().tolist()\n",
    "    KINASES = kinases.to_numpy().flatten().tolist()\n",
    "    for cluster in list(sorted(leiden_df['leiden'].unique())):\n",
    "        cluster_samples = list(leiden_df[leiden_df['leiden'] == cluster].index.values)\n",
    "        for id in cluster_samples:\n",
    "            new = pd.DataFrame([[tf, p] for p in intermediates[id].dropna() for tf in TFs[id].dropna()], columns=['target', 'source'])\n",
    "            edges_df = pd.concat([edges_df, new])\n",
    "            new = pd.DataFrame([[p, kinase] for kinase in kinases[id].dropna()[:rank_threshold] for p in intermediates[id].dropna()], columns=['target', 'source'])\n",
    "            edges_df = pd.concat([edges_df, new])\n",
    "\n",
    "        edges_df = edges_df.groupby(['target', 'source']).size().reset_index()\n",
    "        edges_df.rename(columns={0: \"count\"}, inplace=True)\n",
    "        edges_df['label'] = edges_df['count']\n",
    "        edges_df = edges_df.loc[edges_df.source != edges_df.target]\n",
    "\n",
    "        thresh = edges_df.sort_values(by='count', ascending=False).iloc[n_edges]['count']\n",
    "        if (thresh * 4) <= len(cluster_samples):\n",
    "            display(Markdown(f'No common pathways were identified for at least 25% of samples in cluster {cluster}.'))\n",
    "            continue\n",
    "        edges_show = edges_df.loc[edges_df['count'] >= thresh]\n",
    "        \n",
    "        nodes = pd.unique(edges_show[['target', 'source']].to_numpy().flatten().tolist())\n",
    "        nodes = nodes[~pd.isnull(nodes)]\n",
    "        nodes_df = pd.DataFrame({'id': nodes})\n",
    "        nodes_df['type'] = 'intermediate'\n",
    "        nodes_df['color'] = \"#DCDCDC\"\n",
    "        nodes_df.loc[nodes_df['id'].isin(TRANSCRIPTION_FACTORS), 'type'] = 'TF'\n",
    "        nodes_df.loc[nodes_df['id'].isin(KINASES), 'type'] = 'kinase'\n",
    "        nodes_df.loc[nodes_df['type'] == 'TF', 'color'] = tf_col\n",
    "        nodes_df.loc[nodes_df['type'] == 'kinase', 'color'] = kinase_col\n",
    "        nodes_df['label'] = nodes_df['id']\n",
    "\n",
    "        # Create a graph\n",
    "        G = nx.Graph()\n",
    "\n",
    "        # Add nodes\n",
    "        for idx, row in nodes_df.iterrows():\n",
    "            G.add_node(row['id'], group=row['type'], color=row['color'])\n",
    "\n",
    "        # Add edges\n",
    "        for idx, row in edges_show.iterrows():\n",
    "            G.add_edge(row['source'], row['target'])\n",
    "\n",
    "        # Manually set positions\n",
    "        pos = {}\n",
    "        group_positions = {'kinase': 0.8, 'intermediate': 0.5, 'TF': 0.2} \n",
    "        group_nodes = {group: [] for group in group_positions.keys()}\n",
    "        for node, data in G.nodes(data=True):\n",
    "            group_nodes[data['group']].append(node)\n",
    "        for group, nodes in group_nodes.items():\n",
    "            x_positions = np.linspace(0, 1, len(nodes))\n",
    "            for x, node in zip(x_positions, nodes):\n",
    "                pos[node] = (x, group_positions[group])\n",
    "\n",
    "        nodes_data =G.nodes(data=True)\n",
    "        clus_pathway =  {'kinases':[], 'intermediates': [], 'TFs': []}\n",
    "        for n in nodes_data:\n",
    "            if n[1]['group'] == 'kinase':\n",
    "                clus_pathway['kinases'].append(n[0])\n",
    "            elif n[1]['group'] == 'intermediate':\n",
    "                clus_pathway['intermediates'].append(n[0])\n",
    "            elif n[1]['group'] == 'TF':\n",
    "                clus_pathway['TFs'].append(n[0])\n",
    "        pathway_results[f\"Common X2K Pathways - Downregulated Genes - Cluster {cluster}\"] = convert_to_string(clus_pathway)\n",
    "        \n",
    "        node_colors = [G.nodes[n]['color'] for n in G.nodes()]\n",
    "\n",
    "        if len(G.nodes()) > 60:\n",
    "            display(Markdown(f'Too many nodes were identified for {cluster}.'))\n",
    "            continue\n",
    "\n",
    "        if len(G.nodes()) > 50:\n",
    "            node_size = 600\n",
    "            font_size = 4\n",
    "        elif len(G.nodes()) > 40:\n",
    "            node_size = 400\n",
    "            font_size = 4\n",
    "        else:\n",
    "            node_size = 1500\n",
    "            font_size = 8\n",
    "        plt.figure(figsize=(12, 5))\n",
    "        nx.draw_networkx_nodes(G, pos, node_color=node_colors, node_size=node_size)\n",
    "        nx.draw_networkx_edges(G, pos, edgelist=edges_show.to_records(index=False), width=1)\n",
    "        nx.draw_networkx_labels(G, pos, font_size=font_size, font_family='sans-serif') \n",
    "        plt.savefig(f'figures/{dir}_x2k_pathways_C{cluster}.png', dpi=300)\n",
    "        plt.savefig(f'figures/{dir}_x2k_pathways_C{cluster}.svg', dpi=300)\n",
    "        nx.write_adjlist(G, f'results/{dir}_x2k_pathways_C{cluster}.adjlist')\n",
    "        plt.show()\n",
    "        display(Markdown(f\"__Fig. {fig_counter}{ascii_uppercase[letter_counter]}__ Visualization of pathways commonly identified from up-genes from cluster {cluster}.\"))\n",
    "        letter_counter += 1\n",
    "        display(FileLink(f'figures/{dir}_x2k_pathways_C{cluster}.png', result_html_prefix='Pathways Inferred from Down-Genes PNG: '))\n",
    "        display(FileLink(f'figures/{dir}_x2k_pathways_C{cluster}.svg', result_html_prefix='Pathways Inferred from Down-Genes SVG: '))\n",
    "        display(FileLink(f'results/{dir}_x2k_pathways_C{cluster}.adjlist', result_html_prefix='Download Adjacency List: '))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:\n",
    "    display(Markdown(f\"Common X2K pathways were inferred from down-regulated genes for each cluster and visualized as network graphs (Fig. {fig_counter}{ascii_uppercase[0]}-{ascii_uppercase[letter_counter -1]}).\"))\n",
    "    fig_counter += 1\n",
    "    letter_counter = 0"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:\n",
    "    display(Markdown(create_results_text_prompt(pathway_results, 'These are common pathways identified with X2K(RNA-seq expression to kinases) for up and down regulated genes per cluster. Please write in paragraph form')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file:\n",
    "    display(Markdown(f\"*\\*This section was generated with GPT-4 and should be interpreted with caution*\\*\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if rna_file.value != '' and tumor_validate != [] %}\n",
    "### Validation of Cancer Targets\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file and len(tumor_validate) > 0:\n",
    "    depmap = pd.read_csv('https://minio.dev.maayanlab.cloud/x2k-tr/DepMap_CRISPRGeneEffect.csv.gz', index_col=[0,1], compression='gzip')\n",
    "    depmap_ct = depmap[depmap.index.get_level_values('OncotreePrimaryDisease').isin(tumor_validate)].mean().to_frame('Mean Gene Effect')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file and len(tumor_validate) > 0:\n",
    "    # DepMap Gene Effect scores\n",
    "    flat_list = []\n",
    "    for cluster, bgs in targets.items():\n",
    "        for bg, bg_targets in bgs.items():\n",
    "            for target in bg_targets[:10]:\n",
    "                flat_list.append((target, (cluster, bg)))\n",
    "    df = pd.DataFrame(flat_list, columns=['Target', f'Ranked top 10 in (cluster, background)'])\n",
    "    df = df.groupby('Target').agg(list)\n",
    "    df = pd.merge(df, depmap_ct, left_index=True, right_index=True, how='left')\n",
    "    # Literature occurrences with 'cancer'\n",
    "    generic_occ = []\n",
    "    for target, row in df.iterrows():\n",
    "        generic_occ.append(get_co_occurrence_count([target, 'cancer']))\n",
    "    df['Literature co-occurrence with cancer'] = generic_occ\n",
    "\n",
    "    # Literature occurrences with specific disease\n",
    "    for tumortype in tumor_validate:\n",
    "        specific_occ = []\n",
    "        for target, row in df.iterrows():\n",
    "            specific_occ.append(get_co_occurrence_count([target, tumortype]))\n",
    "        df[f'Literature co-occurrence with {tumortype}'] = specific_occ\n",
    "\n",
    "    display(df)\n",
    "    display(Markdown(f\"__Table {table_counter}__ Top 10 targets from each cluster and background with DepMap gene effect scores and literature co-occurrences with cancer and specific tumor types.\"))\n",
    "    df.to_csv('results/target_validation.csv')\n",
    "    display(FileLink(f\"results/target_validation.csv\", result_html_prefix=' Download DepMap and Literature Validation of Gene Targets:'))\n",
    "    table_counter += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file and len(tumor_validate) > 0:\n",
    "    # DepMap Gene Effect scores\n",
    "    up_df = x2k_up.mean(axis=1).to_frame('Up-Genes MeanRank')\n",
    "    dn_df = x2k_dn.mean(axis=1).to_frame('Dn-Genes MeanRank')\n",
    "    df = pd.concat([up_df, dn_df], axis=1, ignore_index=False)\n",
    "    top_n_up = up_df.sort_values(by='Up-Genes MeanRank').index[:10]\n",
    "    top_n_dn = dn_df.sort_values(by='Dn-Genes MeanRank').index[:10]\n",
    "    df = df.loc[top_n_up.union(top_n_dn)]\n",
    "    df = pd.merge(df, depmap_ct, left_index=True, right_index=True, how='left')\n",
    "    # Literature occurrences with 'cancer'\n",
    "    generic_occ = []\n",
    "    for kinase, row in df.iterrows():\n",
    "        generic_occ.append(get_co_occurrence_count([kinase, 'cancer']))\n",
    "    df['Literature co-occurrence with cancer'] = generic_occ\n",
    "    # Literature occurrences with specific disease\n",
    "    for tumortype in tumor_validate:\n",
    "        specific_occ = []\n",
    "        for kinase, row in df.iterrows():\n",
    "            specific_occ.append(get_co_occurrence_count([kinase, tumortype]))\n",
    "        df[f'Literature co-occurrence with {tumortype}'] = specific_occ\n",
    "    display(df)\n",
    "    display(Markdown(f\"__Table {table_counter}__ DepMap gene effect scores and literature co-occurrences with cancer and specific tumor types for top 10 kinases identified through either the up-regulated and down-regulated genes.\"))\n",
    "    df.to_csv('results/kinase_validation.csv')\n",
    "    display(FileLink(f\"results/kinase_validation.csv\", result_html_prefix=' Download DepMap and Literature Validation of Kinases:'))\n",
    "    table_counter += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if phospho_file.value != '' and impute_phospho_expr.raw_value %}\n",
    "if has_phospho_file:\n",
    "    percentages = []\n",
    "    idx = []\n",
    "    i = 0\n",
    "    for _, row in tqdm(phospho_df.iterrows(), leave=False, mininterval=30, total=len(phospho_df)):\n",
    "        value_counts = row.astype(float).isna().value_counts()\n",
    "        pr = value_counts[0]/len(row)\n",
    "        if pr > percent_rows_phospho:\n",
    "            idx.append(i)\n",
    "        percentages.append(pr)\n",
    "        i += 1\n",
    "    ax = pd.Series(percentages).hist()\n",
    "    ax.set_ylabel(\"# Phosphosites\")\n",
    "    ax.set_xlabel(\"% with values\")\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if phospho_file.value != '' and impute_phospho_expr.raw_value %}\n",
    "if has_phospho_file:\n",
    "    display(Markdown(f\"__Fig. {fig_counter}__ Distribution of missing values in phosphoproteomic data.\"))\n",
    "    fig_counter += 1\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "if has_phospho_file:\n",
    "    phospho_df.replace([np.inf, -np.inf], np.nan, inplace=True)\n",
    "    {% if phospho_file.value != '' and impute_phospho_expr.raw_value %}\n",
    "    n1 = len(phospho_df)\n",
    "    phospho_df_imputed = phospho_df.iloc[idx].T.fillna(phospho_df.iloc[idx].mean(axis=1)).T.dropna()\n",
    "    zscored_phospho = zscore(phospho_df_imputed, axis=1)\n",
    "    zscored_samps = zscore(phospho_df_imputed, axis=0)\n",
    "    display(Markdown(f\"Dropped {n1 - len(phospho_df_imputed)} rows with missing values. The distribution of missing values in phosphoproteomic data was visualized in Figure {fig_counter - 1}.\"))\n",
    "    {% elif phospho_file.value != '' and not impute_phospho_expr.raw_value %}\n",
    "    phospho_df.dropna(inplace=True)\n",
    "    zscored_phospho = zscore(phospho_df, axis=1)\n",
    "    zscored_samps = zscore(phospho_df, axis=0)\n",
    "    {% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if phospho_file.value != '' %}\n",
    "#### Most Variable Phosphosites\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_phospho_file:\n",
    "    zscored_phospho_var = zscored_samps.var(axis=1).sort_values(ascending=False)\n",
    "    zscored_phospho_var_list = zscored_phospho_var.index[:100].tolist()\n",
    "    df = pd.DataFrame(zscored_phospho_var, columns=['Variance'])\n",
    "    df.index.names = ['Phosphosite']\n",
    "    if '.' in zscored_phospho_var.index.values[0]:\n",
    "        df['Gene Symbol'] = df.index.map(lambda x: lookup(x.split('.')[0]) if lookup(x.split('.')[0]) else x.split('.')[0])\n",
    "    elif '_' in zscored_phospho_var.index.values[0]:\n",
    "        df['Gene Symbol'] = df.index.map(lambda x: lookup(x.split('_')[0]) if lookup(x.split('_')[0]) else x.split('_')[0])\n",
    "    else:\n",
    "        df['Gene Symbol'] = df.index.map(lambda x: lookup(x) if lookup(x) else x)\n",
    "    display(df.head(10))\n",
    "    display(Markdown(f\"__Table {table_counter}__ Top 10 most variable phosphosites.\"))\n",
    "    table_counter += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_phospho_file:    \n",
    "    try:\n",
    "        g = create_annotated_clustermap(zscored_phospho.loc[zscored_phospho_var_list], meta_df, leiden_df, default_meta_cols, cbar_label='Z-Score', ylabel='Samples', xlabel='Phosphosites')\n",
    "    except:\n",
    "        plt.clf()\n",
    "        g = create_clustermap(zscored_phospho.loc[zscored_phospho_var_list], cbar_label='Z-Score', ylabel='Samples', xlabel='Phosphosites')\n",
    "    clustergrammer_res = clustergrammer_link(zscored_phospho.loc[zscored_phospho_var_list], f'{phospho_filename}_top_var_phosphosites')\n",
    "    if clustergrammer_res:\n",
    "        display(Markdown(f\"[View in Clustergrammer]({clustergrammer_res})\"))\n",
    "    ax = g.ax_heatmap\n",
    "    ax.set_ylabel('Top 100 Phosphosites by Variance')\n",
    "    ax.set_xlabel('Samples')\n",
    "    g.savefig(f'figures/{phospho_filename}_top_var_phosphosites_heatmap.png', dpi=300)\n",
    "    g.savefig(f'figures/{phospho_filename}_top_var_phosphosites_heatmap.svg', dpi=300)\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_phospho_file:\n",
    "    display(Markdown(f\"__Fig. {fig_counter}__ Hierarchically clustered heatmap of the top 100 most variable phosphosites.\"))\n",
    "    display(FileLink(f'figures/{phospho_filename}_top_var_phosphosites_heatmap.png', result_html_prefix='Most Variable Phosphosites Heatmap PNG: '))\n",
    "    display(FileLink(f'figures/{phospho_filename}_top_var_phosphosites_heatmap.svg', result_html_prefix='Most Variable Phosphosites Heatmap SVG: '))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_phospho_file:\n",
    "    display(Markdown(f\"The top 10 most variable phosphosites were visualized as a hierarchically clustered heatmap (Fig. {fig_counter}).\"))\n",
    "    fig_counter += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file and has_phospho_file:  \n",
    "    adata = sc.AnnData(zscored_samps.T.values, dtype=\"float\")\n",
    "    adata.var['phosphosite_names'] = zscored_samps.index.values\n",
    "    adata.obs['samples'] = zscored_samps.columns.values\n",
    "    adata.var['var_rank'] = (-np.var(adata.X, axis=0, dtype=\"float\")).argsort()\n",
    "    adata = adata[:, adata.var.var_rank < 5000]\n",
    "\n",
    "    adata.obs['leiden'] = adata.obs['samples'].map(lambda s: leiden_df.loc[s]['leiden'] if s in leiden_df.index.values else 'N/A' )\n",
    "    sc.pp.pca(adata)\n",
    "    sc.pp.neighbors(adata, n_neighbors=n_neighbors)  # create neighborhood graph\n",
    "    sc.tl.umap(adata, min_dist=min_dist, alpha=0.1) \n",
    "    clusters = adata.obs['leiden'].unique()\n",
    "    cmap = plt.cm.get_cmap('tab20b', len(clusters))\n",
    "    for i, cluster in enumerate(sorted(clusters)):\n",
    "        mask = adata.obs['leiden'] == cluster\n",
    "        ax = plt.scatter(x=adata.obsm['X_umap'][mask, 0], y=adata.obsm['X_umap'][mask, 1], \n",
    "                    color=cmap(i), label=f'Cluster {cluster}', s=10)\n",
    "    plt.legend(title='Leiden clusters', bbox_to_anchor=(1.05, 1), loc='best')\n",
    "    ax.axes.set_xticks([])\n",
    "    ax.axes.set_yticks([])\n",
    "    plt.xlabel('UMAP-1')\n",
    "    plt.ylabel('UMAP-2')\n",
    "    plt.savefig(f'figures/{phospho_filename}_umap_leiden_rna_clusters.png', dpi=300, bbox_inches='tight')\n",
    "    plt.savefig(f'figures/{phospho_filename}_umap_leiden_rna_clusters.svg', dpi=300, bbox_inches='tight')\n",
    "elif has_phospho_file:\n",
    "    adata = sc.AnnData(zscored_samps.T.values, dtype=\"int\")\n",
    "    adata.var['protein_names'] = zscored_samps.index.values\n",
    "    adata.obs['samples'] = zscored_samps.columns.values\n",
    "    # Sort proteins by variance\n",
    "    adata.var['var_rank'] = (-np.var(adata.X, axis=0, dtype=\"float\")).argsort()\n",
    "    adata = adata[:, adata.var.var_rank < 5000]\n",
    "    # UMAP\n",
    "    sc.pp.pca(adata)\n",
    "    sc.pp.neighbors(adata, n_neighbors=n_neighbors)  # create neighborhood graph\n",
    "    sc.tl.umap(adata, min_dist=min_dist, alpha=0.1)  # embed umap based on neighborhood graph\n",
    "    sc.tl.leiden(adata, resolution=resolution)  # clustering\n",
    "    clusters = adata.obs['leiden'].unique()\n",
    "    cmap = plt.cm.get_cmap('tab20b', len(clusters))\n",
    "    for i, cluster in enumerate(sorted(clusters)):\n",
    "        mask = adata.obs['leiden'] == cluster\n",
    "        ax = plt.scatter(x=adata.obsm['X_umap'][mask, 0], y=adata.obsm['X_umap'][mask, 1], \n",
    "                    color=cmap(i), label=f'Cluster {cluster}', s=10)\n",
    "    plt.legend(title='Leiden clusters', bbox_to_anchor=(1.05, 1), loc='best')\n",
    "    ax.axes.set_xticks([])\n",
    "    ax.axes.set_yticks([])\n",
    "    plt.xlabel('UMAP-1')\n",
    "    plt.ylabel('UMAP-2')\n",
    "    plt.savefig(f'figures/{phospho_filename}_umap_leiden.png', dpi=300, bbox_inches='tight')\n",
    "    plt.savefig(f'figures/{phospho_filename}_umap_leiden.svg', dpi=300, bbox_inches='tight')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_phospho_file and has_rna_file:\n",
    "    display(Markdown(f\"__Fig. {fig_counter}__ UMAP visualization of top 5000 most variable phosphosites colored by leiden clusters determined from RNA-seq expression.\"))\n",
    "    display(FileLink(f'figures/{phospho_filename}_umap_leiden_rna_clusters.png', result_html_prefix='Phosphosite UMAP colored by RNA Leiden clusters PNG: '))\n",
    "    display(FileLink(f'figures/{phospho_filename}_umap_leiden_rna_clusters.svg', result_html_prefix='Phosphosite UMAP colored by RNA Leiden clusters SVG: '))\n",
    "elif has_phospho_file:\n",
    "    display(Markdown(f\"__Fig. {fig_counter}__ Leiden clusters determined from Z-scored phosphosites for 5000 most variable proteins.\"))\n",
    "    display(FileLink(f'figures/{phospho_filename}_umap_leiden.png', result_html_prefix='Leiden clusters UMAP PNG: '))\n",
    "    display(FileLink(f'figures/{phospho_filename}_umap_leiden.svg', result_html_prefix='Leiden clusters UMAP SVG: '))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_phospho_file and has_rna_file:\n",
    "    display(Markdown(f\"A UMAP visualization was created of top 5000 most variable phosphosites colored by leiden clusters determined from RNA-seq expression (Fig. {fig_counter}).\"))\n",
    "    fig_counter += 1\n",
    "elif has_phospho_file:\n",
    "    display(Markdown(f\"A UMAP visualization was created of top 5000 most variable phosphosites colored by leiden clusters (Fig. {fig_counter}).\"))\n",
    "    fig_counter += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_phospho_file:\n",
    "    kea_res = {}\n",
    "    for patient in tqdm(zscored_phospho.columns, leave=False, mininterval=30, desc='Computing KEA3 Scores'):\n",
    "        kea_res[patient] = {}\n",
    "        if '.' in zscored_phospho.index.values[0]:\n",
    "            down_rows = list(map(lambda g: lookup(g.split('.')[0]), filter(lambda g: lookup(g.split('.')[0]) , zscored_samps[patient].astype(float).sort_values(ascending=True).index.values)))[:kea3_n]\n",
    "            up_rows = list(map(lambda g: lookup(g.split('.')[0]), filter(lambda g: lookup(g.split('.')[0]) , zscored_samps[patient].astype(float).sort_values(ascending=False).index.values)))[:kea3_n]\n",
    "        elif '_' in zscored_phospho.index.values[0]:\n",
    "            down_rows = list(map(lambda g: lookup(g.split('_')[0]), filter(lambda g: lookup(g.split('_')[0]) , zscored_samps[patient].astype(float).sort_values(ascending=True).index.values)))[:kea3_n]\n",
    "            up_rows = list(map(lambda g: lookup(g.split('_')[0]), filter(lambda g: lookup(g.split('_')[0]) , zscored_samps[patient].astype(float).sort_values(ascending=False).index.values)))[:kea3_n]\n",
    "        else:\n",
    "            down_rows = list(map(lambda g: lookup(g.split('_')[0]), filter(lambda g: lookup(g) , zscored_samps[patient].astype(float).sort_values(ascending=True).index.values)))[:kea3_n]\n",
    "            up_rows = list(map(lambda g: lookup(g.split('_')[0]), filter(lambda g: lookup(g) , zscored_samps[patient].astype(float).sort_values(ascending=False).index.values)))[:kea3_n]\n",
    "        \n",
    "        up_kinases = get_kea3_results(down_rows, f'{patient}_up_phosphosites')\n",
    "        down_kinases = get_kea3_results(up_rows, f'{patient}_down_phosphosites')\n",
    "        kea_res[patient]['up'] = up_kinases\n",
    "        kea_res[patient]['down'] = down_kinases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if phospho_file.value != '' %}\n",
    "#### Full KEA3 Heatmap - Up Phosphosites\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_phospho_file:    \n",
    "    up_kea_comb = pd.concat([kea_res[p]['up'] for p in kea_res], axis=1)\n",
    "    up_kea_comb.columns = list(kea_res.keys())\n",
    "    try:\n",
    "        g = create_annotated_clustermap(up_kea_comb.astype(float), meta_df, leiden_df, default_meta_cols, cbar_label='Mean Rank', ylabel='Kinases', xlabel='Samples')\n",
    "    except:\n",
    "        g = create_clustermap(up_kea_comb.astype(float), cbar_label='Mean Rank', ylabel='Kinases', xlabel='Samples')\n",
    "    clustergrammer_res = clustergrammer_link(up_kea_comb.astype(float), f'{phospho_filename}_kea_up')\n",
    "    if clustergrammer_res:\n",
    "        display(Markdown(f\"[View in Clustergrammer]({clustergrammer_res})\"))\n",
    "    g.savefig(f'figures/{phospho_filename}_kea_up_heatmap.png', dpi=300)\n",
    "    g.savefig(f'figures/{phospho_filename}_kea_up_heatmap.svg', dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_phospho_file:\n",
    "    display(Markdown(f\"__Fig. {fig_counter}{ascii_uppercase[letter_counter]}__ Integrated Mean Rank from KEA3 from up-regulated phosphosites for all samples and all kinases based on z-scored phosphorylation.\"))\n",
    "    letter_counter += 1\n",
    "    display(FileLink(f'figures/{phospho_filename}_kea_up_heatmap.png', result_html_prefix='KEA3 Up Heatmap PNG: '))\n",
    "    display(FileLink(f'figures/{phospho_filename}_kea_up_heatmap.svg', result_html_prefix='KEA3 Up Heatmap SVG: '))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if phospho_file.value != '' %}\n",
    "#### KEA3 Heatmap Top 20 Variable Kinases - Up Phosphosites\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_phospho_file:   \n",
    "    top_20_var_kinases = up_kea_comb.var(axis=1).sort_values(ascending=False).index[0:20]\n",
    "    try:\n",
    "        g = create_annotated_clustermap(up_kea_comb.loc[top_20_var_kinases].astype(float), meta_df, leiden_df, default_meta_cols, cbar_label='Mean Rank', ylabel='Top 20 Kinases by Variance', yticklabels=True)\n",
    "    except:\n",
    "        plt.clf()\n",
    "        g = create_clustermap(up_kea_comb.loc[top_20_var_kinases].astype(float), cbar_label='Mean Rank', ylabel='Top 20 Kinases by Variance', yticklabels=True)\n",
    "    clustergrammer_res = clustergrammer_link(up_kea_comb.loc[top_20_var_kinases].astype(float), f'{phospho_filename}_kea_up_top20')\n",
    "    if clustergrammer_res:\n",
    "        display(Markdown(f\"[View in Clustergrammer]({clustergrammer_res})\"))\n",
    "    g.savefig(f'figures/{phospho_filename}_kea_up_top_var_heatmap.png', dpi=300)\n",
    "    g.savefig(f'figures/{phospho_filename}_kea_up_top_var_heatmap.svg', dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_phospho_file:\n",
    "    display(Markdown(f\"__Fig. {fig_counter}{ascii_uppercase[letter_counter]}__ Integrated Mean Rank from KEA3 from up-regulated phosphosites for all samples and top 20 most variable kinases based on z-scored phosphorylation.\"))\n",
    "    letter_counter += 1\n",
    "    display(FileLink(f'figures/{phospho_filename}_kea_up_top_var_heatmap.png', result_html_prefix='KEA3 Up Top 20 Variable Kinases Heatmap PNG: '))\n",
    "    display(FileLink(f'figures/{phospho_filename}_kea_up_top_var_heatmap.svg', result_html_prefix='KEA3 Up Top 20 Variable Kinases Heatmap SVG: '))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if phospho_file.value != '' %}\n",
    "#### Full KEA3 Heatmap - Down Phosphosites\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_phospho_file:    \n",
    "    dn_kea_comb = pd.concat([kea_res[p]['down'] for p in kea_res], axis=1)\n",
    "    dn_kea_comb.columns = list(kea_res.keys())\n",
    "    try:\n",
    "        g = create_annotated_clustermap(dn_kea_comb.astype(float), meta_df, leiden_df, default_meta_cols, cbar_label='Mean Rank', ylabel='Kinases', xlabel='Samples')\n",
    "    except:\n",
    "        plt.clf()\n",
    "        g = create_clustermap(dn_kea_comb.astype(float), cbar_label='Mean Rank', ylabel='Kinases', xlabel='Samples')\n",
    "    clustergrammer_res = clustergrammer_link(dn_kea_comb.astype(float), f'{phospho_filename}_kea_dn')\n",
    "    if clustergrammer_res:\n",
    "        display(Markdown(f\"[View in Clustergrammer]({clustergrammer_res})\"))\n",
    "    g.savefig(f'figures/{phospho_filename}_kea_dn_heatmap.png', dpi=300)\n",
    "    g.savefig(f'figures/{phospho_filename}_kea_dn_heatmap.svg', dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_phospho_file:    \n",
    "    display(Markdown(f\"__Fig. {fig_counter}{ascii_uppercase[letter_counter]}__ Integrated Mean Rank from KEA3 from down-regulated phosphosites for all samples and all kinases based on z-scored phosphorylation.\"))\n",
    "    letter_counter += 1\n",
    "    display(FileLink(f'figures/{phospho_filename}_kea_dn_heatmap.png', result_html_prefix='KEA3 Down Heatmap PNG: '))\n",
    "    display(FileLink(f'figures/{phospho_filename}_kea_dn_heatmap.svg', result_html_prefix='KEA3 Down Heatmap SVG: '))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if phospho_file.value != '' %}\n",
    "#### KEA3 Heatmap Top 20 Variable Kinases - Down Phosphosites\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_phospho_file:\n",
    "    top_20_var_kinases = dn_kea_comb.var(axis=1).sort_values(ascending=False).index[0:20]\n",
    "    try:\n",
    "        g = create_annotated_clustermap(dn_kea_comb.loc[top_20_var_kinases].astype(float), meta_df, leiden_df, default_meta_cols, cbar_label='Mean Rank', ylabel='Top 20 Kinases by Variance', yticklabels=True)\n",
    "    except:\n",
    "        plt.clf()\n",
    "        g = create_clustermap(dn_kea_comb.loc[top_20_var_kinases].astype(float), cbar_label='Mean Rank', ylabel='Top 20 Kinases by Variance', yticklabels=True)\n",
    "    clustergrammer_res = clustergrammer_link(dn_kea_comb.loc[top_20_var_kinases], f'{phospho_filename}_kea_dn_top20')\n",
    "    if clustergrammer_res:\n",
    "        display(Markdown(f\"[View in Clustergrammer]({clustergrammer_res})\"))\n",
    "    g.savefig(f'figures/{phospho_filename}_kea_dn_top_var_heatmap.png', dpi=300)\n",
    "    g.savefig(f'figures/{phospho_filename}_kea_dn_top_var_heatmap.svg', dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_phospho_file:    \n",
    "    display(Markdown(f\"__Fig. {fig_counter}{ascii_uppercase[letter_counter]}__ Integrated Mean Rank from KEA3 from down-regulated phosphosites for all samples and top 20 most variable kinases based on z-scored phosphorylation.\"))\n",
    "    display(FileLink(f'figures/{phospho_filename}_kea_dn_top_var_heatmap.png', result_html_prefix='KEA3 Down Top 20 Variable Kinases Heatmap PNG: '))\n",
    "    display(FileLink(f'figures/{phospho_filename}_kea_dn_top_var_heatmap.svg', result_html_prefix='KEA3 Down Top 20 Variable Kinases Heatmap SVG: '))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_phospho_file:\n",
    "    display(Markdown(f'KEA3 results were generated for the top {kea3_n} up and down-regulated phosphosites for each sample z-scored sample. The results were then aggregated and visualized as heatmaps for all kinases as well as the top 20 most variable kinases (Fig. {fig_counter}{ascii_uppercase[0]}-{ascii_uppercase[letter_counter]}).'))\n",
    "    letter_counter = 0 \n",
    "    fig_counter += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if rna_file.value != '' and phospho_file.value != '' %}\n",
    "### Transcriptomics + Phosphoproteomics Analysis Results\n",
    "#### Kinases enriched in both X2K and phospho- pipelines\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file and has_phospho_file:\n",
    "    rank_threshold = 30\n",
    "\n",
    "    rec_kinases = {}\n",
    "    rec_kinases_shuffled = {}\n",
    "    shuffled_ids = list(kea_res.keys())\n",
    "    np.random.shuffle(shuffled_ids)\n",
    "    for i, id in enumerate(kea_res):\n",
    "        if id not in x2k_up:\n",
    "            continue\n",
    "        up = x2k_up[id].astype(float).sort_values(ascending=True).index.to_list()\n",
    "        up_phospho = kea_res[id]['up'].astype(float).sort_values(by='Score', ascending=True).index.to_list()\n",
    "        up_phospho_rand = kea_res[shuffled_ids[i]]['up'].astype(float).sort_values(by='Score', ascending=True).index.to_list()\n",
    "        rec_kinases[id] = list(set(up[:rank_threshold]) & set(up_phospho[:rank_threshold]))\n",
    "        rec_kinases_shuffled[id] = list(set(up[:rank_threshold]) & set(up_phospho_rand[:rank_threshold]))\n",
    "    up_up_df_shuffled = pd.DataFrame.from_dict(rec_kinases_shuffled, orient='index').T\n",
    "    up_up_df = pd.DataFrame.from_dict(rec_kinases, orient='index').T\n",
    "    up_up_df.to_csv(f\"results/recovered_up_phospho_up_genes.csv\", index=False)\n",
    "\n",
    "    rec_kinases = {}\n",
    "    rec_kinases_shuffled = {}\n",
    "    shuffled_ids = list(kea_res.keys())\n",
    "    np.random.shuffle(shuffled_ids)\n",
    "    for i, id  in enumerate(kea_res):\n",
    "        if id not in x2k_up:\n",
    "            continue\n",
    "        dn = x2k_dn[id].astype(float).sort_values(ascending=True).index.to_list()\n",
    "        dn_phospho = kea_res[id]['down'].astype(float).sort_values(by='Score', ascending=True).index.to_list()\n",
    "        dn_phospho_rand = kea_res[shuffled_ids[i]]['down'].astype(float).sort_values(by='Score', ascending=True).index.to_list()\n",
    "        rec_kinases[id] = list(set(dn[:rank_threshold]) & set(dn_phospho[:rank_threshold]))\n",
    "        rec_kinases_shuffled[id] = list(set(dn[:rank_threshold]) & set(dn_phospho_rand[:rank_threshold]))\n",
    "    dn_dn_df_shuffled = pd.DataFrame.from_dict(rec_kinases_shuffled, orient='index').T\n",
    "    dn_dn_df = pd.DataFrame.from_dict(rec_kinases, orient='index').T\n",
    "    dn_dn_df.to_csv(f\"results/recovered_dn_phospho_dn_genes.csv\", index=False)\n",
    "\n",
    "    rec_kinases = {}\n",
    "    rec_kinases_shuffled = {}\n",
    "    shuffled_ids = list(kea_res.keys())\n",
    "    np.random.shuffle(shuffled_ids)\n",
    "    for i, id in enumerate(kea_res):\n",
    "        if id not in x2k_up:\n",
    "            continue\n",
    "        dn = x2k_dn[id].astype(float).sort_values(ascending=True).index.to_list()\n",
    "        up_phospho = kea_res[id]['up'].astype(float).sort_values(by='Score', ascending=True).index.to_list()\n",
    "        up_phospho_rand = kea_res[shuffled_ids[i]]['up'].astype(float).sort_values(by='Score', ascending=True).index.to_list()\n",
    "        rec_kinases[id] = list(set(dn[:rank_threshold]) & set(up_phospho[:rank_threshold]))\n",
    "        rec_kinases_shuffled[id] = list(set(dn[:rank_threshold]) & set(up_phospho_rand[:rank_threshold]))\n",
    "    up_dn_df_shuffled = pd.DataFrame.from_dict(rec_kinases_shuffled, orient='index').T\n",
    "    up_dn_df = pd.DataFrame.from_dict(rec_kinases, orient='index').T\n",
    "    up_dn_df.to_csv(f\"results/recovered_up_phospho_dn_genes.csv\", index=False)\n",
    "\n",
    "    rec_kinases = {}\n",
    "    rec_kinases_shuffled = {}\n",
    "    shuffled_ids = list(kea_res.keys())\n",
    "    np.random.shuffle(shuffled_ids)\n",
    "    for i, id in enumerate(kea_res):\n",
    "        if id not in x2k_up:\n",
    "            continue\n",
    "        up = x2k_up[id].astype(float).sort_values(ascending=True).index.to_list()\n",
    "        dn_phospho = kea_res[id]['down'].astype(float).sort_values(by='Score', ascending=True).index.to_list()\n",
    "        dn_phospho_rand = kea_res[shuffled_ids[i]]['down'].astype(float).sort_values(by='Score', ascending=True).index.to_list()\n",
    "        rec_kinases[id] = list(set(up[:rank_threshold]) & set(dn_phospho[:rank_threshold]))\n",
    "        rec_kinases_shuffled[id] = list(set(up[:rank_threshold]) & set(dn_phospho_rand[:rank_threshold]))\n",
    "    dn_up_df_shuffled = pd.DataFrame.from_dict(rec_kinases_shuffled, orient='index').T\n",
    "    dn_up_df = pd.DataFrame.from_dict(rec_kinases, orient='index').T\n",
    "    dn_up_df.to_csv(f\"results/recovered_dn_phospho_up_genes.csv\", index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file and has_phospho_file:   \n",
    "    display(up_up_df.head())\n",
    "    display(Markdown(f\"__Table {table_counter}__ Kinases recovered from both the X2K pipeline and the phosphoproteomic data for up-genes and up-phosphosites.\"))\n",
    "    table_counter += 1\n",
    "    display(FileLink(f\"results/recovered_up_phospho_up_genes.csv\", result_html_prefix='Download Recovered Kinases (up-up):'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file and has_phospho_file:      \n",
    "    display(dn_dn_df.head())\n",
    "    display(Markdown(f\"__Table {table_counter}__ Kinases recovered from both the X2K pipeline and the phosphoproteomic data for down-genes and down-phosphosites.\"))\n",
    "    table_counter += 1\n",
    "    display(FileLink(f\"results/recovered_dn_phospho_dn_genes.csv\", result_html_prefix='Download Recovered Kinases (down-down):'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file and has_phospho_file:      \n",
    "    display(up_dn_df.head())\n",
    "    display(Markdown(f\"__Table {table_counter}__ Kinases recovered from both the X2K pipeline and the phosphoproteomic data for up-genes and down-phosphosites.\"))\n",
    "    table_counter += 1\n",
    "    display(FileLink(f\"results/recovered_up_phospho_dn_genes.csv\", result_html_prefix='Download Recovered Kinases (up-down):'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file and has_phospho_file:\n",
    "    display(dn_up_df.head())\n",
    "    display(Markdown(f\"__Table {table_counter}__ Kinases recovered from both the X2K pipeline and the phosphoproteomic data for down-genes and up-phosphosites.\"))\n",
    "    table_counter += 1\n",
    "    display(FileLink(f\"results/recovered_dn_phospho_up_genes.csv\", result_html_prefix='Download Recovered Kinases (down-up):'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file and has_phospho_file:\n",
    "    up_up = pd.read_csv(\"results/recovered_up_phospho_up_genes.csv\")\n",
    "    up_dn = pd.read_csv(\"results/recovered_up_phospho_dn_genes.csv\")\n",
    "    dn_up = pd.read_csv(\"results/recovered_dn_phospho_dn_genes.csv\")\n",
    "    dn_dn = pd.read_csv(\"results/recovered_dn_phospho_dn_genes.csv\")\n",
    "\n",
    "    samples_to_plot = []\n",
    "    for cluster in sorted(leiden_df['leiden'].unique()):\n",
    "        cluster_samples = list(set(leiden_df[leiden_df['leiden'] == cluster].index.values).intersection(up_up.columns))\n",
    "        dat = []\n",
    "        dat_shuffled = []\n",
    "        for id in cluster_samples:\n",
    "            dat.append([id, 'Up', 'Up', up_up.shape[0]-sum(up_up[id].isna())])\n",
    "            dat.append([id, 'Up', 'Down', up_dn.shape[0]-sum(up_dn[id].isna())])\n",
    "            dat.append([id, 'Down', 'Down', dn_dn.shape[0]-sum(dn_dn[id].isna())])\n",
    "            dat.append([id, 'Down', 'Up', dn_up.shape[0]-sum(dn_up[id].isna())])\n",
    "            dat_shuffled.append([id, 'Up', 'Up', up_up_df_shuffled.shape[0]-sum(up_up_df_shuffled[id].isna())])\n",
    "            dat_shuffled.append([id, 'Up', 'Down', up_dn_df_shuffled.shape[0]-sum(up_dn_df_shuffled[id].isna())])\n",
    "            dat_shuffled.append([id, 'Down', 'Down', dn_dn_df_shuffled.shape[0]-sum(dn_dn_df_shuffled[id].isna())])\n",
    "            dat_shuffled.append([id, 'Down', 'Up', dn_up_df_shuffled.shape[0]-sum(dn_up_df_shuffled[id].isna())])\n",
    "\n",
    "        dat = pd.DataFrame(dat, columns=[\"id\", \"Kinases\", \"Genes\", \"Overlap\"])\n",
    "        dat = dat.pivot(index=['id'], columns=['Kinases', 'Genes'], values='Overlap')\n",
    "        dat_shuffled = pd.DataFrame(dat_shuffled, columns=[\"id\", \"Kinases\", \"Genes\", \"Overlap\"])\n",
    "        sorted_samples = dat.sum(axis=1).sort_values(ascending=False).index.values\n",
    "        dat = dat.loc[sorted_samples]\n",
    "        dat_shuffled = dat_shuffled.pivot(index=['id'], columns=['Kinases', 'Genes'], values='Overlap')\n",
    "        dat.loc['random'] = dat_shuffled.loc[sorted_samples].mean()\n",
    "\n",
    "        # Identify Samples with better than average recovery\n",
    "        samp_p_values = [ttest_ind(dat.loc[samp], dat.loc['random'], alternative='greater') for samp in dat.drop('random').index.values]\n",
    "        g = dat.plot(kind='bar', stacked=True, color=['coral', 'peachpuff', 'royalblue', 'lightblue'], fontsize=8, figsize=(len(up_up.columns)*0.15, 6))\n",
    "        bars = g.patches\n",
    "        top_bars_idx = (len(bars) // 4) -1\n",
    "        for idx, bar in enumerate(bars):\n",
    "            if ((idx + 1) % (len(bars) // 4)) == 0:\n",
    "                bar.set_hatch('/')\n",
    "        for i, bar in enumerate(g.containers[3][:-1]):\n",
    "            if samp_p_values[i].pvalue < 0.0005:\n",
    "                samples_to_plot.append((dat.index.values[i], samp_p_values[i].pvalue))\n",
    "                plt.text(bar.xy[0] +(bar.get_width() / 4), bar.xy[1] + bar.get_height(), \"***\", fontsize=16)\n",
    "            elif samp_p_values[i].pvalue < 0.005:\n",
    "                samples_to_plot.append((dat.index.values[i], samp_p_values[i].pvalue))\n",
    "                plt.text(bar.xy[0] +(bar.get_width() / 4), bar.xy[1] + bar.get_height(), \"**\", fontsize=16)\n",
    "            elif samp_p_values[i].pvalue < 0.05:\n",
    "                samples_to_plot.append((dat.index.values[i], samp_p_values[i].pvalue))\n",
    "                plt.text(bar.xy[0] +(bar.get_width() / 4), bar.xy[1] + bar.get_height(), \"*\", fontsize=16)\n",
    "\n",
    "            \n",
    "        g.set_ylabel(\"Kinase Overlap\")\n",
    "        g.set_xlabel(\"Samples\")\n",
    "        #g.legend(title=\"Kinases,Genes\", loc='best')\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(f\"figures/kinase_recovery_c{cluster}.svg\", dpi=300)\n",
    "        plt.savefig(f\"figures/kinase_recovery_c{cluster}.png\", dpi=300)\n",
    "        plt.show()\n",
    "        display(Markdown(f\"__Fig. {fig_counter}{ascii_uppercase[letter_counter]}__ Overlap of kinases recovered from both the X2K pipeline and the phosphoproteomic data for up/down genes and phosphosites per patient in cluster {cluster}.\"))\n",
    "        letter_counter += 1\n",
    "        display(FileLink(f\"figures/kinase_recovery_c{cluster}.png\", result_html_prefix=f'Barchart of Kinase Recovery Cluster {cluster} PNG: '))\n",
    "\n",
    "        display(FileLink(f\"figures/kinase_recovery_c{cluster}.svg\", result_html_prefix=f'Barchart of Kinase Recovery Cluster {cluster} SVG: '))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file and has_phospho_file:\n",
    "    display(Markdown(f\"The recovery of kinases from both the X2K pipeline and the phosphoproteomic data for both up and down genes and phosphosites per patient in each cluster is compared to random recovery and visualized as a stacked bar chart, which each color matching a pair of directions of gene and phosphosite regulation (Fig. {fig_counter}{ascii_uppercase[0]}-{ascii_uppercase[letter_counter - 1]}).\"))\n",
    "    letter_counter = 0\n",
    "    fig_counter += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if prot_file.value != '' and impute_protein_expr.raw_value %}\n",
    "if has_prot_file:\n",
    "    percentages = []\n",
    "    idx = []\n",
    "    i = 0\n",
    "    for _, row in tqdm(prot_df.iterrows(), total=len(prot_df)):\n",
    "        value_counts = row.astype(float).isna().value_counts()\n",
    "        pr = value_counts[0]/len(row)\n",
    "        if pr > percent_rows_protein:\n",
    "            idx.append(i)\n",
    "        percentages.append(pr)\n",
    "        i += 1\n",
    "    ax = pd.Series(percentages).hist()\n",
    "    ax.set_ylabel(\"# Proteins\")\n",
    "    ax.set_xlabel(\"% with values\")\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "{% if prot_file.value != '' and impute_protein_expr.raw_value %}\n",
    "if has_prot_file:\n",
    "    display(Markdown(f\"__Fig. {fig_counter}__ Distribution of missing values in proteomic data.\"))\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter code_exec\n",
    "if has_prot_file:\n",
    "    {% if impute_protein_expr.raw_value %}\n",
    "    n1 = len(prot_df)\n",
    "    prot_df_imputed = prot_df.iloc[idx].T.fillna(prot_df.iloc[idx].mean(axis=1)).T.astype(float)\n",
    "    zscored_prot = zscore(prot_df_imputed, axis=1)\n",
    "    zscored_samps = zscore(prot_df_imputed, axis=0)\n",
    "    display(Markdown(f\"Dropped {n1 - len(prot_df_imputed)} rows with missing values while imputing with averages. Distribution of missing values in proteomics data is displayed in Figure {fig_counter}.\"))\n",
    "    fig_counter += 1\n",
    "    {% else %}\n",
    "    prot_df.dropna(inplace=True)\n",
    "    zscored_prot = zscore(prot_df, axis=1)\n",
    "    zscored_samps = zscore(prot_df, axis=0)\n",
    "    {% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_prot_file:\n",
    "    top_var_proteins = zscored_samps.var(axis=1).sort_values(ascending=False)\n",
    "    top_var_proteins_list = top_var_proteins.index[:100].tolist()\n",
    "    df = pd.DataFrame(top_var_proteins, columns=['Variance'])\n",
    "    df.index.names = ['Protein']\n",
    "    display(df.head(10))\n",
    "    display(Markdown(f\"__Table {table_counter}__ Top 10 most variable proteins based on z-scored expression.\"))\n",
    "    table_counter += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if prot_file.value != '' %}\n",
    "### Top 100 Proteins by Variance\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_prot_file:\n",
    "    try:\n",
    "        g = create_annotated_clustermap(zscored_prot.loc[top_var_proteins_list], meta_df, leiden_df, default_meta_cols, cbar_label='Z-Score', xlabel='Samples', ylabel='Top 100 Proteins by Variance')\n",
    "    except:\n",
    "        plt.clf()\n",
    "        g = create_clustermap(zscored_prot.loc[top_var_proteins_list], cbar_label='Z-Score', xlabel='Samples', ylabel='Top 100 Proteins by Variance')\n",
    "    clustergrammer_res = clustergrammer_link(zscored_prot.loc[top_var_proteins_list], f'{prot_filename}_zscored_top_var_proteins')\n",
    "    if clustergrammer_res:\n",
    "        display(Markdown(f\"[View in Clustergrammer]({clustergrammer_res})\"))\n",
    "    g.savefig(f'figures/{prot_filename}_top_var_proteins_heatmap.png', dpi=300)\n",
    "    g.savefig(f'figures/{prot_filename}_top_var_proteins_heatmap.svg', dpi=300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_prot_file:\n",
    "    display(Markdown(f\"__Fig. {fig_counter}__ Hierarchically clustered heatmap of the top 100 most variable proteins.\"))\n",
    "    display(FileLink(f'figures/{prot_filename}_top_var_proteins_heatmap.png', result_html_prefix='Most Variable Proteins Heatmap PNG: '))\n",
    "    display(FileLink(f'figures/{prot_filename}_top_var_proteins_heatmap.svg', result_html_prefix='Most Variable Proteins Heatmap SVG: '))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_prot_file:\n",
    "    display(Markdown(f\"Top 100 most variable proteins were identified based on Z-scored expression and visualized (Fig. {fig_counter}).\"))\n",
    "    fig_counter += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file and has_prot_file:  \n",
    "    adata = sc.AnnData(zscored_samps.T.values, dtype=\"float\")\n",
    "    adata.var['protein_names'] = zscored_samps.index.values\n",
    "    adata.obs['samples'] = zscored_samps.columns.values\n",
    "    adata.var['var_rank'] = (-np.var(adata.X, axis=0, dtype=\"float\")).argsort()\n",
    "    adata = adata[:, adata.var.var_rank < 5000]\n",
    "\n",
    "    adata.obs['leiden'] = adata.obs['samples'].map(lambda s: leiden_df.loc[s]['leiden'] if s in leiden_df.index.values else 'N/A' )\n",
    "    sc.pp.pca(adata)\n",
    "    sc.pp.neighbors(adata, n_neighbors=n_neighbors)  # create neighborhood graph\n",
    "    sc.tl.umap(adata, min_dist=min_dist, alpha=0.1) \n",
    "    clusters = adata.obs['leiden'].unique()\n",
    "    cmap = plt.cm.get_cmap('tab20b', len(clusters))\n",
    "    for i, cluster in enumerate(sorted(clusters)):\n",
    "        mask = adata.obs['leiden'] == cluster\n",
    "        ax = plt.scatter(x=adata.obsm['X_umap'][mask, 0], y=adata.obsm['X_umap'][mask, 1], \n",
    "                    color=cmap(i), label=f'Cluster {cluster}', s=10)\n",
    "    plt.legend(title='Leiden clusters', bbox_to_anchor=(1.05, 1), loc='best')\n",
    "    ax.axes.set_xticks([])\n",
    "    ax.axes.set_yticks([])\n",
    "    plt.xlabel('UMAP-1')\n",
    "    plt.ylabel('UMAP-2')\n",
    "    plt.savefig(f'figures/{prot_filename}_umap_leiden_rna_clusters.png', dpi=300, bbox_inches='tight')\n",
    "    plt.savefig(f'figures/{prot_filename}_umap_leiden_rna_clusters.svg', dpi=300, bbox_inches='tight')\n",
    "elif has_prot_file:\n",
    "    adata = sc.AnnData(zscored_samps.T.values, dtype=\"int\")\n",
    "    adata.var['protein_names'] = zscored_samps.index.values\n",
    "    adata.obs['samples'] = zscored_samps.columns.values\n",
    "    # Sort proteins by variance\n",
    "    adata.var['var_rank'] = (-np.var(adata.X, axis=0, dtype=\"float\")).argsort()\n",
    "    adata = adata[:, adata.var.var_rank < 5000]\n",
    "\n",
    "    # UMAP\n",
    "    sc.pp.pca(adata)\n",
    "    sc.pp.neighbors(adata, n_neighbors=n_neighbors)  # create neighborhood graph\n",
    "    sc.tl.umap(adata, min_dist=min_dist, alpha=0.1)  # embed umap based on neighborhood graph\n",
    "    sc.tl.leiden(adata, resolution=resolution)  # clustering\n",
    "    clusters = adata.obs['leiden'].unique()\n",
    "    cmap = plt.cm.get_cmap('tab20b', len(clusters))\n",
    "    for i, cluster in enumerate(sorted(clusters)):\n",
    "        mask = adata.obs['leiden'] == cluster\n",
    "        ax = plt.scatter(x=adata.obsm['X_umap'][mask, 0], y=adata.obsm['X_umap'][mask, 1], \n",
    "                    color=cmap(i), label=f'Cluster {cluster}', s=10)\n",
    "    plt.legend(title='Leiden clusters', bbox_to_anchor=(1.05, 1), loc='best')\n",
    "    ax.axes.set_xticks([])\n",
    "    ax.axes.set_yticks([])\n",
    "    plt.xlabel('UMAP-1')\n",
    "    plt.ylabel('UMAP-2')\n",
    "    plt.savefig(f'figures/{prot_filename}_umap_leiden.png', dpi=300, bbox_inches='tight')\n",
    "    plt.savefig(f'figures/{prot_filename}_umap_leiden.svg', dpi=300, bbox_inches='tight')\n",
    "    display(Markdown(f\"__Fig. {fig_counter}__ Leiden clusters determined from Z-scored proteomic expression for 5000 most variable proteins.\"))\n",
    "    display(FileLink(f'figures/{prot_filename}_umap_leiden.png', result_html_prefix='Leiden clusters UMAP PNG: '))\n",
    "    display(FileLink(f'figures/{prot_filename}_umap_leiden.svg', result_html_prefix='Leiden clusters UMAP SVG: '))\n",
    "    fig_counter += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file and has_prot_file:\n",
    "    display(Markdown(f\"__Fig. {fig_counter}__ UMAP visualization of top 5000 most variable proteins colored by leiden clusters determined from RNA-seq expression.\"))\n",
    "    display(FileLink(f'figures/{prot_filename}_umap_leiden_rna_clusters.png', result_html_prefix='Protein UMAP colored by RNA Leiden clusters PNG: '))\n",
    "    display(FileLink(f'figures/{prot_filename}_umap_leiden_rna_clusters.svg', result_html_prefix='Protein UMAP colored by RNA Leiden clusters SVG: '))\n",
    "elif has_prot_file:\n",
    "    display(Markdown(f\"__Fig. {fig_counter}__ Leiden clusters determined from Z-scored proteomic expression for 5000 most variable proteins.\"))\n",
    "    display(FileLink(f'figures/{prot_filename}_umap_leiden.png', result_html_prefix='Leiden clusters UMAP PNG: '))\n",
    "    display(FileLink(f'figures/{prot_filename}_umap_leiden.svg', result_html_prefix='Leiden clusters UMAP SVG: '))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file and has_prot_file:\n",
    "    display(Markdown(f\"Proteomics expression was visualized using UMAP and colored by RNA-seq determined leiden clusters (Fig. {fig_counter}).\"))\n",
    "    fig_counter += 1\n",
    "elif has_prot_file:\n",
    "    display(Markdown(f\"Proteomics expression was visualized using UMAP and leiden clusters were determined for the top 5000 most variable proteins (Fig. {fig_counter}).\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if prot_file.value != '' and rna_file.value != '' %}\n",
    "### Annotated Cell-Surface Targets with Proteomic Expression\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file and has_prot_file:\n",
    "    cmap = {\"None\":YlGnBu(0), \"ARCHS4\": YlGnBu(1), \"GTEx\":YlGnBu(2), \"TS\": YlGnBu(4), \"ARCHS4-GTEx\":YlGnBu(3),  \"ARCHS4-TS\": YlGnBu(5), \"GTEx-TS\": YlGnBu(6), \"All\": YlGnBu(7)}\n",
    "    annot_data = []\n",
    "    for t in membrane_target_mat.index:\n",
    "        if t in zscored_prot.index:\n",
    "            row = []\n",
    "            for cluster in membrane_target_mat.columns:\n",
    "                c = cluster.split(' ')[1]\n",
    "                cluster_samples = list(leiden_df[leiden_df['leiden'] == c].index.values)\n",
    "                cluster_samples = list(set(cluster_samples).intersection(zscored_prot.columns))\n",
    "                clus_mean = (zscored_samps.loc[t][cluster_samples]).mean()\n",
    "                row.append(str(clus_mean)[0:4])\n",
    "            annot_data.append(row)\n",
    "        else:\n",
    "            annot_data.append(['NA']*membrane_target_mat.shape[1])\n",
    "    h = membrane_target_mat.shape[0]\n",
    "    g = sns.clustermap(membrane_target_mat, annot=annot_data, fmt='', figsize=(4,(0.3*h+2*(h<15))), cmap=YlGnBu, cbar_pos=None, dendrogram_ratio=0.1-(h<40)*0.01*(h-30), row_cluster=False, xticklabels=True, yticklabels=True, annot_kws={\"fontsize\":8})\n",
    "    g.ax_row_dendrogram.legend(handles=[Rectangle((0, 0), 0, 0, color=val, label=key) for key, val in cmap.items()],\n",
    "                                    title='Background', loc='upper right')\n",
    "\n",
    "    plt.show()\n",
    "\n",
    "    g.savefig(f'figures/{rna_filename}_prot_membrane_target_matrix.png', dpi=300)\n",
    "    g.savefig(f'figures/{rna_filename}_prot_membrane_target_matrix.svg', dpi=300)\n",
    "    membrane_target_mat.to_csv(f'results/{rna_filename}_prot_membrane_target_mat.csv')\n",
    "    display(Markdown(f\"__Fig. {fig_counter}__ Membrane target matrix annotated with average z-scored expression from proteomic data.\"))\n",
    "    display(FileLink(f'results/{rna_filename}_prot_membrane_target_mat.csv', result_html_prefix='Download Membrane Target Matrix: '))\n",
    "    display(FileLink(f'figures/{rna_filename}_prot_membrane_target_matrix.png', result_html_prefix='Membrane Target Matrix PNG: '))\n",
    "    display(FileLink(f'figures/{rna_filename}_prot_membrane_target_matrix.svg', result_html_prefix='Membrane Target Matrix SVG: '))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file and has_prot_file:\n",
    "    display(Markdown(f\"Membrane target matrix was annotated with average Z-scored expression across samples comprising each cluster using the uploaded proteomics data (Fig. {fig_counter}).\"))\n",
    "    fig_counter += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%appyter markdown\n",
    "{% if prot_file.value != '' and rna_file.value != '' and phospho_file.value != '' %}\n",
    "### Recovered Kinases Proteomic Expression per Patient\n",
    "{% endif %}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file and has_prot_file and has_phospho_file:\n",
    "    bar_width = .30\n",
    "    sig_kinase_proteins = []\n",
    "    for s, pval in samples_to_plot:\n",
    "        up_up_dict = up_up_df.to_dict(orient='list')\n",
    "        dn_dn_dict = dn_dn_df.to_dict(orient='list')\n",
    "        up_dn_dict = up_dn_df.to_dict(orient='list')\n",
    "        dn_up_dict = dn_up_df.to_dict(orient='list')\n",
    "        rec_kinases_sample = list(set(up_up_dict[s] + dn_dn_dict[s] + up_dn_dict[s] + dn_up_dict[s]))\n",
    "        idx_overlap = list(set(rec_kinases_sample).intersection(zscored_samps.index))\n",
    "        prot_vals = zscored_samps[s].loc[idx_overlap].sort_values(ascending=False)\n",
    "        prot_vals_avg = zscored_samps.loc[prot_vals.index].mean(axis=1)\n",
    "        y_err_std = zscored_samps.loc[prot_vals.index].std(axis=1).values\n",
    "        r1 = np.arange(len(prot_vals))\n",
    "        r2 = [x + bar_width for x in r1]\n",
    "        ax1 = plt.figure(figsize=(10, 6))\n",
    "        ax2 = plt.bar(r1, prot_vals, width=bar_width, color='lightblue', label='Sample Protein Expression', ecolor='#a8a8a8', capsize=3)\n",
    "        plt.bar(r2, prot_vals_avg, width=bar_width, color='grey', label='Mean Protein Expression', yerr=y_err_std, ecolor='#a8a8a8', capsize=3)\n",
    "        for i, b in enumerate(ax2.patches):\n",
    "            if prot_vals.iloc[i] > prot_vals_avg.iloc[i] +  abs((2 * y_err_std[i])):\n",
    "                sig_kinase_proteins.append((prot_vals.index[i], 'up' if prot_vals.iloc[i] > prot_vals_avg.iloc[i] else 'down', s))\n",
    "                plt.text(b.get_x() + b.get_width() / 2, b.get_height(), '*', ha='center', va='center', fontsize=16)\n",
    "            elif prot_vals.iloc[i] < prot_vals_avg.iloc[i] - abs((2 * y_err_std[i])):\n",
    "                sig_kinase_proteins.append((prot_vals.index[i], 'up' if prot_vals.iloc[i] > prot_vals_avg.iloc[i] else 'down', s))\n",
    "                plt.text(b.get_x() + b.get_width() / 2, b.get_height(), '*', ha='center', va='center', fontsize=16)\n",
    "        plt.xticks([r + bar_width for r in range(len(prot_vals.index))], prot_vals.index, rotation=90)\n",
    "        plt.legend()\n",
    "        plt.xlabel(f'Kinases Recovered for {s}')\n",
    "        plt.ylabel('Z-Scored Protein Expression')\n",
    "        plt.savefig(f'figures/{prot_filename}_prot_kinase_expr_{s}.png', dpi=300, bbox_inches='tight')\n",
    "        plt.savefig(f'figures/{prot_filename}_prot_kinase_expr_{s}.svg', dpi=300, bbox_inches='tight')\n",
    "        plt.show()\n",
    "        display(Markdown(f\"__Fig. {fig_counter}{ascii_uppercase[letter_counter]}__ Z-scored protein expression of overlapping kinases from X2K from up genes and directly from phosphoproteomics through KEA3 from up phosphosites.\"))\n",
    "        letter_counter += 1 \n",
    "        display(FileLink(f'figures/{prot_filename}_prot_kinase_expr_{s}.png', result_html_prefix=f'Z-scored protein expression of overlapping kinases from X2K and KEA3 for {s} PNG: '))\n",
    "        display(FileLink(f'figures/{prot_filename}_prot_kinase_expr_{s}.svg', result_html_prefix=f'Z-scored protein expression of overlapping kinases from X2K and KEA3 for {s} SVG: '))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file and has_prot_file and has_phospho_file:\n",
    "    display(Markdown(f\"Z-scored protein expression for kinases identified for patients that showed significantly better than random recovery of kinases in X2K when compared to the kinase enrichment from the phosphoproteomics data (Fig. {fig_counter}{ascii_uppercase[0]}-{ascii_uppercase[letter_counter]}).\"))\n",
    "    letter_counter = 0\n",
    "    fig_counter += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file and has_prot_file and has_phospho_file:\n",
    "    display(Markdown(create_results_text_prompt(sig_kinase_proteins, 'The following kinases were found to have significantly different protein expression compared to the average expression of the proteins across samples. This result was computed for samples that had greater than average recovery of kinases from X2K (RNA-seq expression to kinases) pipeline compared to kinases directly inferred from phosphoproteomics.')))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if has_rna_file and has_prot_file and has_phospho_file:\n",
    "    display(Markdown(f\"*\\*This section was generated with GPT-4 and should be interpreted with caution*\\*\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Discussion\n",
    "\n",
    "While the X2K pipeline predicted many kinases and TFs with well-studied individual roles in cancer, the signaling pathways and regulatory networks connecting are less well known. By inferring co-expressed intermediates and identifying kinases and TFs enriched in the same tumor sample, the X2K analysis presents possible novel regulatory pathways linking upstream kinases to their downstream targets.\n",
    "\n",
    "The target identification component of the pipeline also facilitates hypothesis generation regarding novel cancer targets. Many predicted targets are not directly linked to the specific cancer subtype, but rather connected to other cancers, immune processes, or mechanisms related to cell proliferation and cancer cell survival."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Conclusions\n",
    "\n",
    "Finding targets and key to cell signaling pathways that are uniquely activated in tumor cells while absent from all normal tissues and cell types remains a significant challenge. In addition, targets identified from analyzing transcriptomics do not ensure that the targets are also expressed at the protein level. Furthermore, heterogeneity within the tumor, across patients of the same cancer type, and across cancers, underscore the need for subtype-specific personalized target identification. Our analysis and workflows address some of these challenges by combining phosphoproteomics, proteomics, and transcriptomics with comprehensive normal tissue and cell type atlases to identify safer targets for cancer subtypes."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## References\n",
    "\n",
    "[1] Chen EY, Xu H, Gordonov S, Lim MP, Perkins MH, Ma'ayan A. Expression2Kinases: mRNA profiling linked to multiple upstream regulatory layers. Bioinformatics. 2012 Jan 1;28(1):105-11. doi: 10.1093/bioinformatics/btr625. Epub 2011 Nov 10.\n",
    "\n",
    "[2] Clarke DJB, Kuleshov MV, Schilder BM, Torre D, Duffy ME, Keenan AB, Lachmann A, Feldmann AS, Gundersen GW, Silverstein MC, Wang Z, Ma'ayan A. eXpression2Kinases (X2K) Web: linking expression signatures to upstream cell signaling networks. Nucleic Acids Res. 2018 Jul 2;46(W1):W171-W179. doi: 10.1093/nar/gky458.\n",
    "\n",
    "[2] Marino GB, Ngai M, Clarke DJB, Fleishman RH, Deng EZ, Xie Z, Ahmed N, Ma'ayan A. GeneRanger and TargetRanger: processed gene and protein expression levels across cells and tissues for target discovery. Nucleic Acids Res. 2023 Jul 5;51(W1):W213-W224. doi: 10.1093/nar/gkad399.\n",
    "\n",
    "[4] Cancer Genome Atlas Research Network; Van Allen EM, Cherniack AD, Ciriello G, Sander C, Schultz N. Oncogenic Signaling Pathways in The Cancer Genome Atlas. Cell. 2018 Apr 5;173(2):321-337.e10. doi: 10.1016/j.cell.2018.03.035. PMID: 29625050; PMCID: PMC6070353.\n",
    "\n",
    "[5] Keenan AB, Torre D, Lachmann A, Leong AK, Wojciechowicz ML, Utti V, Jagodnik KM, Kropiwnicki E, Wang Z, Ma'ayan A. ChEA3: transcription factor enrichment analysis by orthogonal omics integration. Nucleic Acids Res. 2019 Jul 2;47(W1):W212-W224. doi: 10.1093/nar/gkz446.\n",
    "\n",
    "[6] Berger SI, Posner JM, Ma'ayan A. Genes2Networks: connecting lists of gene symbols using mammalian protein interactions databases. BMC Bioinformatics. 2007 Oct 4;8:372. doi: 10.1186/1471-2105-8-372.\n",
    "\n",
    "[7] Kuleshov MV, Xie Z, London ABK, Yang J, Evangelista JE, Lachmann A, Shu I, Torre D, Ma'ayan A. KEA3: improved kinase enrichment analysis via data integration. Nucleic Acids Res. 2021 Jul 2;49(W1):W304-W316. doi: 10.1093/nar/gkab359.\n",
    "\n",
    "[8] Ge SX, Son EW, Yao R. iDEP: an integrated web application for differential expression and pathway analysis of RNA-Seq data. BMC Bioinformatics. 2018 Dec 19;19(1):534. doi: 10.1186/s12859-018-2486-6.\n",
    "\n",
    "[9] Reimand J, Isserlin R, Voisin V, Kucera M, Tannus-Lopes C, Rostamianfar A, Wadi L, Meyer M, Wong J, Xu C, Merico D, Bader GD. Pathway enrichment analysis and visualization of omics data using g:Profiler, GSEA, Cytoscape and EnrichmentMap. Nat Protoc. 2019 Feb;14(2):482-517. doi: 10.1038/s41596-018-0103-9.\n",
    "\n",
    "[10] Jin Y, Ratnam K, Chuang PY, Fan Y, Zhong Y, Dai Y, Mazloom AR, Chen EY, D'Agati V, Xiong H, Ross MJ, Chen N, Ma'ayan A, He JC. A systems approach identifies HIPK2 as a key regulator of kidney fibrosis. Nat Med. 2012 Mar 11;18(4):580-8. doi: 10.1038/nm.2685. Erratum in: Nat Med. 2021 Aug;27(8):1483.\n",
    "\n",
    "[11] Bosse KR, Raman P, Zhu Z, Lane M, Martinez D, Heitzeneder S, Rathi KS, Kendsersky NM, Randall M, Donovan L, Morrissy S, Sussman RT, Zhelev DV, Feng Y, Wang Y, Hwang J, Lopez G, Harenza JL, Wei JS, Pawel B, Bhatti T, Santi M, Ganguly A, Khan J, Marra MA, Taylor MD, Dimitrov DS, Mackall CL, Maris JM. Identification of GPC2 as an Oncoprotein and Candidate Immunotherapeutic Target in High-Risk Neuroblastoma. Cancer Cell. 2017 Sep 11;32(3):295-309.e12. doi: 10.1016/j.ccell.2017.08.003.\n",
    "\n",
    "[12] GTEx Consortium. The Genotype-Tissue Expression (GTEx) project. Nat Genet. 2013 Jun;45(6):580-5. doi: 10.1038/ng.2653.\n",
    "\n",
    "[13] Peters C, Brown S. Antibody-drug conjugates as novel anti-cancer chemotherapeutics. Biosci Rep. 2015 Jun 12;35(4):e00225. doi: 10.1042/BSR20150089.\n",
    "\n",
    "[14] Sadelain M, Rivière I, Brentjens R. Targeting tumours with genetically enhanced T lymphocytes. Nat Rev Cancer. 2003 Jan;3(1):35-45. doi: 10.1038/nrc971.\n",
    "\n",
    "[15] Ho WY, Blattman JN, Dossett ML, Yee C, Greenberg PD. Adoptive immunotherapy: engineering T cell responses as biologic weapons for tumor mass destruction. Cancer Cell. 2003 May;3(5):431-7. doi: 10.1016/s1535-6108(03)00113-2.\n",
    "\n",
    "[16] Orentas RJ, Yang JJ, Wen X, Wei JS, Mackall CL, Khan J. Identification of cell surface proteins as potential immunotherapy targets in 12 pediatric cancers. Front Oncol. 2012 Dec 17;2:194. doi: 10.3389/fonc.2012.00194.\n",
    "\n",
    "[17] Lee JK, Bangayan NJ, Chai T, Smith BA, Pariva TE, Yun S, Vashisht A, Zhang Q, Park JW, Corey E, Huang J, Graeber TG, Wohlschlegel J, Witte ON. Systemic surfaceome profiling identifies target antigens for immune-based therapy in subtypes of advanced prostate cancer. Proc Natl Acad Sci U S A. 2018 May 8;115(19):E4473-E4482. doi: 10.1073/pnas.1802354115. Epub 2018 Apr 23.\n",
    "\n",
    "[18] Ferguson ID, Patiño-Escobar B, Tuomivaara ST, Lin YT, Nix MA, Leung KK, Kasap C, Ramos E, Nieves Vasquez W, Talbot A, Hale M, Naik A, Kishishita A, Choudhry P, Lopez-Girona A, Miao W, Wong SW, Wolf JL, Martin TG 3rd, Shah N, Vandenberg S, Prakash S, Besse L, Driessen C, Posey AD Jr, Mullins RD, Eyquem J, Wells JA, Wiita AP. The surfaceome of multiple myeloma cells suggests potential immunotherapeutic strategies and protein markers of drug resistance. Nat Commun. 2022 Jul 15;13(1):4121. doi: 10.1038/s41467-022-31810-6.\n",
    "\n",
    "[19] Lachmann A, Torre D, Keenan AB, Jagodnik KM, Lee HJ, Wang L, Silverstein MC, Ma'ayan A. Massive mining of publicly available RNA-seq data from human and mouse. Nature Communications 9. Article number: 1366 (2018), doi: 10.1038/s41467-018-03751-6.\n",
    "\n",
    "[20] Tabula Sapiens Consortium. The Tabula Sapiens: A multiple-organ, single-cell transcriptomic atlas of humans. Science. 2022 May 13;376(6594):eabl4896. doi: 10.1126/science.abl4896\n",
    "\n",
    "[21] Traag VA, Waltman L, van Eck NJ. From Louvain to Leiden: guaranteeing well-connected communities. Sci Rep. 2019 Mar 26;9(1):5233. doi: 10.1038/s41598-019-41695-z.\n",
    "\n",
    "[22] Ritchie ME, Phipson B, Wu D, Hu Y, Law CW, Shi W, Smyth GK. limma powers differential expression analyses for RNA-sequencing and microarray studies. Nucleic Acids Res. 2015 Apr 20;43(7):e47. doi: 10.1093/nar/gkv007.\n",
    "\n",
    "[23] Chen EY, Tan CM, Kou Y, Duan Q, Wang Z, Meirelles GV, Clark NR, Ma'ayan A. Enrichr: interactive and collaborative HTML5 gene list enrichment analysis tool. BMC Bioinformatics. 2013 Apr 15;14:128. doi: 10.1186/1471-2105-14-128.\n",
    "\n",
    "[24] Ashburner M, Ball CA, Blake JA, Botstein D, Butler H, Cherry JM, Davis AP, Dolinski K, Dwight SS, Eppig JT, Harris MA, Hill DP, Issel-Tarver L, Kasarskis A, Lewis S, Matese JC, Richardson JE, Ringwald M, Rubin GM, Sherlock G. Gene ontology: tool for the unification of biology. The Gene Ontology Consortium. Nat Genet. 2000 May;25(1):25-9. doi: 10.1038/75556.\n",
    "\n",
    "[25] Pico AR, Kelder T, van Iersel MP, Hanspers K, Conklin BR, Evelo C. WikiPathways: pathway editing for the people. PLoS Biol. 2008 Jul 22;6(7):e184. doi: 10.1371/journal.pbio.0060184.\n",
    "\n",
    "[26] Eppig JT, Smith CL, Blake JA, Ringwald M, Kadin JA, Richardson JE, Bult CJ. Mouse Genome Informatics (MGI): Resources for Mining Mouse Genetic, Genomic, and Biological Data in Support of Primary and Translational Research. Methods Mol Biol. 2017;1488:47-73. doi: 10.1007/978-1-4939-6427-7_3.\n",
    "\n",
    "[27]  Binder JX, Pletscher-Frankild S, Tsafou K, Stolte C, O'Donoghue SI, Schneider R, Jensen LJ. COMPARTMENTS: unification and visualization of protein subcellular localization evidence. Database (Oxford). 2014 Feb 25;2014:bau012. doi: 10.1093/database/bau012.\n",
    "\n",
    "[28] Thul PJ, Lindskog C. The human protein atlas: A spatial map of the human proteome. Protein Sci. 2018 Jan;27(1):233-244. doi: 10.1002/pro.3307. Epub 2017 Oct 10.\n",
    "\n",
    "[29] Nusinow DP, Szpyt J, Ghandi M, Rose CM, McDonald ER 3rd, Kalocsay M, Jané-Valbuena J, Gelfand E, Schweppe DK, Jedrychowski M, Golji J, Porter DA, Rejtar T, Wang YK, Kryukov GV, Stegmeier F, Erickson BK, Garraway LA, Sellers WR, Gygi SP. Quantitative Proteomics of the Cancer Cell Line Encyclopedia. Cell. 2020 Jan 23;180(2):387-402.e16. doi: 10.1016/j.cell.2019.12.023.\n",
    "\n",
    "[30] Tsherniak A, Vazquez F, Montgomery PG, Weir BA, Kryukov G, Cowley GS, Gill S, Harrington WF, Pantel S, Krill-Burger JM, Meyers RM, Ali L, Goodale A, Lee Y, Jiang G, Hsiao J, Gerath WFJ, Howell S, Merkel E, Ghandi M, Garraway LA, Root DE, Golub TR, Boehm JS, Hahn WC. Defining a Cancer Dependency Map. Cell. 2017 Jul 27;170(3):564-576.e16. doi: 10.1016/j.cell.2017.06.010."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
